{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Multi-omics analysis for cancer",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/rtsundby/blog/blob/master/Multi_omics_analysis_for_cancer.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JaJiRqWCQWIE"
      },
      "source": [
        "# Exploring how to use multi-omics data to understand the cancer phenotype\n",
        "\n",
        "This notebook was created for a live demo of the cptac python data API. You are welcome to play along at home. If you want to do that, enter the following URL in your browser: https://cutt.ly/ekMMht7\n",
        "\n",
        "Your next step will be to save this notebook as a copy so that you can edit it. (File->Save a copy in Drive). Now you can run the code below and modify it if you want to explore you own questions."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_bA6g_jCOu5u",
        "outputId": "a33faf2f-86f1-410b-dde3-28a66dc3fa4d"
      },
      "source": [
        "#server call to set up the VM, only necessary on Google Colabs\n",
        "!pip install cptac"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting cptac\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/79/4d/a9585475aa7fc80eee2933f1a11ff4b261ac91781ec456664c0f3fa2921d/cptac-0.8.8-py3-none-any.whl (5.5MB)\n",
            "\u001b[K     |████████████████████████████████| 5.5MB 3.9MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy>=1.16.3 in /usr/local/lib/python3.6/dist-packages (from cptac) (1.19.5)\n",
            "Collecting beautifulsoup4>=4.7.1\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/d1/41/e6495bd7d3781cee623ce23ea6ac73282a373088fcd0ddc809a047b18eae/beautifulsoup4-4.9.3-py3-none-any.whl (115kB)\n",
            "\u001b[K     |████████████████████████████████| 122kB 49.3MB/s \n",
            "\u001b[?25hRequirement already satisfied: requests>=2.21.0 in /usr/local/lib/python3.6/dist-packages (from cptac) (2.23.0)\n",
            "Requirement already satisfied: urllib3>=1.24.2 in /usr/local/lib/python3.6/dist-packages (from cptac) (1.24.3)\n",
            "Requirement already satisfied: statsmodels>=0.10.0 in /usr/local/lib/python3.6/dist-packages (from cptac) (0.10.2)\n",
            "Requirement already satisfied: scipy>=1.2.1 in /usr/local/lib/python3.6/dist-packages (from cptac) (1.4.1)\n",
            "Collecting xlrd==1.2.0\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/b0/16/63576a1a001752e34bf8ea62e367997530dc553b689356b9879339cf45a4/xlrd-1.2.0-py2.py3-none-any.whl (103kB)\n",
            "\u001b[K     |████████████████████████████████| 112kB 51.6MB/s \n",
            "\u001b[?25hRequirement already satisfied: pandas>=0.25.1 in /usr/local/lib/python3.6/dist-packages (from cptac) (1.1.5)\n",
            "Requirement already satisfied: packaging>=19.2 in /usr/local/lib/python3.6/dist-packages (from cptac) (20.9)\n",
            "Collecting openpyxl>=2.6.0\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/d4/c5/1a5f82b3020bfb27f21b302f96c8ae6a34475070015d1b1e0b197a97e2af/openpyxl-3.0.6-py2.py3-none-any.whl (242kB)\n",
            "\u001b[K     |████████████████████████████████| 245kB 39.6MB/s \n",
            "\u001b[?25hCollecting soupsieve>1.2; python_version >= \"3.0\"\n",
            "  Downloading https://files.pythonhosted.org/packages/41/e7/3617a4b988ed7744743fb0dbba5aa0a6e3f95a9557b43f8c4740d296b48a/soupsieve-2.2-py3-none-any.whl\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.6/dist-packages (from requests>=2.21.0->cptac) (2.10)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.6/dist-packages (from requests>=2.21.0->cptac) (3.0.4)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.6/dist-packages (from requests>=2.21.0->cptac) (2020.12.5)\n",
            "Requirement already satisfied: patsy>=0.4.0 in /usr/local/lib/python3.6/dist-packages (from statsmodels>=0.10.0->cptac) (0.5.1)\n",
            "Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.6/dist-packages (from pandas>=0.25.1->cptac) (2.8.1)\n",
            "Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.6/dist-packages (from pandas>=0.25.1->cptac) (2018.9)\n",
            "Requirement already satisfied: pyparsing>=2.0.2 in /usr/local/lib/python3.6/dist-packages (from packaging>=19.2->cptac) (2.4.7)\n",
            "Requirement already satisfied: jdcal in /usr/local/lib/python3.6/dist-packages (from openpyxl>=2.6.0->cptac) (1.4.1)\n",
            "Requirement already satisfied: et-xmlfile in /usr/local/lib/python3.6/dist-packages (from openpyxl>=2.6.0->cptac) (1.0.1)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.6/dist-packages (from patsy>=0.4.0->statsmodels>=0.10.0->cptac) (1.15.0)\n",
            "Installing collected packages: soupsieve, beautifulsoup4, xlrd, openpyxl, cptac\n",
            "  Found existing installation: beautifulsoup4 4.6.3\n",
            "    Uninstalling beautifulsoup4-4.6.3:\n",
            "      Successfully uninstalled beautifulsoup4-4.6.3\n",
            "  Found existing installation: xlrd 1.1.0\n",
            "    Uninstalling xlrd-1.1.0:\n",
            "      Successfully uninstalled xlrd-1.1.0\n",
            "  Found existing installation: openpyxl 2.5.9\n",
            "    Uninstalling openpyxl-2.5.9:\n",
            "      Successfully uninstalled openpyxl-2.5.9\n",
            "Successfully installed beautifulsoup4-4.9.3 cptac-0.8.8 openpyxl-3.0.6 soupsieve-2.2 xlrd-1.2.0\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "epu85fBiYPCW"
      },
      "source": [
        "# Part 1. setup up and intial exploration\n",
        "\n",
        "In the code cell below are the three basic steps for setting up the package\n",
        "1. Import the package, and the utility functions\n",
        "2. Download data (this is only for the first use of the package)\n",
        "3. Instantiate a variable for the cancer type of interest\n",
        "\n",
        "If you have questions about the basic set up, please visit out tutorials at https://paynelab.github.io/cptac/"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zI6IsM1YQlB6",
        "outputId": "ca377e14-a8b2-466c-cbe8-740156cf2a1e"
      },
      "source": [
        "import cptac\n",
        "from cptac import utils\n",
        "cptac.download(\"endometrial\")\n",
        "endo = cptac.Endometrial()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.\n",
            "  import pandas.util.testing as tm\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            ""
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "J-xe9LewYzQK"
      },
      "source": [
        "Now we are ready to grab our data. The package gives you datasets in a tabular/matrix format called a *dataframe*. In this analysis, we are going to work with genomic, proteomic and clinical data. Often we will use a package function to merge multiple datatypes together into one table for convenience.\n",
        "\n",
        "Our first step is find the most commonly mutated genes in endometrial cancer. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 491
        },
        "id": "71SpyeUZYM0h",
        "outputId": "1304caf2-98f2-4ee1-958b-3f5d457b7d30"
      },
      "source": [
        "mutations = utils.get_frequently_mutated(endo)\n",
        "sorted_mutations = mutations.sort_values(by=[\"Missense_Mut\"], ascending=False)\n",
        "sorted_mutations.head(15)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Name</th>\n",
              "      <th>Gene</th>\n",
              "      <th>Unique_Samples_Mut</th>\n",
              "      <th>Missense_Mut</th>\n",
              "      <th>Truncation_Mut</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>157</th>\n",
              "      <td>PIK3CA</td>\n",
              "      <td>0.494737</td>\n",
              "      <td>0.484211</td>\n",
              "      <td>0.010526</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>166</th>\n",
              "      <td>PTEN</td>\n",
              "      <td>0.789474</td>\n",
              "      <td>0.463158</td>\n",
              "      <td>0.568421</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>116</th>\n",
              "      <td>KRAS</td>\n",
              "      <td>0.326316</td>\n",
              "      <td>0.326316</td>\n",
              "      <td>0.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>55</th>\n",
              "      <td>CTNNB1</td>\n",
              "      <td>0.305263</td>\n",
              "      <td>0.305263</td>\n",
              "      <td>0.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>158</th>\n",
              "      <td>PIK3R1</td>\n",
              "      <td>0.389474</td>\n",
              "      <td>0.231579</td>\n",
              "      <td>0.189474</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>134</th>\n",
              "      <td>MUC16</td>\n",
              "      <td>0.189474</td>\n",
              "      <td>0.178947</td>\n",
              "      <td>0.052632</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>149</th>\n",
              "      <td>OBSCN</td>\n",
              "      <td>0.157895</td>\n",
              "      <td>0.157895</td>\n",
              "      <td>0.042105</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>98</th>\n",
              "      <td>HUWE1</td>\n",
              "      <td>0.157895</td>\n",
              "      <td>0.157895</td>\n",
              "      <td>0.031579</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>96</th>\n",
              "      <td>HERC2</td>\n",
              "      <td>0.178947</td>\n",
              "      <td>0.157895</td>\n",
              "      <td>0.063158</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>87</th>\n",
              "      <td>FBXW7</td>\n",
              "      <td>0.189474</td>\n",
              "      <td>0.157895</td>\n",
              "      <td>0.042105</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>205</th>\n",
              "      <td>TP53</td>\n",
              "      <td>0.221053</td>\n",
              "      <td>0.157895</td>\n",
              "      <td>0.073684</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>223</th>\n",
              "      <td>ZFHX4</td>\n",
              "      <td>0.200000</td>\n",
              "      <td>0.147368</td>\n",
              "      <td>0.094737</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>65</th>\n",
              "      <td>DNAH10</td>\n",
              "      <td>0.147368</td>\n",
              "      <td>0.147368</td>\n",
              "      <td>0.031579</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>194</th>\n",
              "      <td>SYNE1</td>\n",
              "      <td>0.168421</td>\n",
              "      <td>0.147368</td>\n",
              "      <td>0.063158</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>136</th>\n",
              "      <td>MUC5B</td>\n",
              "      <td>0.147368</td>\n",
              "      <td>0.136842</td>\n",
              "      <td>0.031579</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "Name    Gene  Unique_Samples_Mut  Missense_Mut  Truncation_Mut\n",
              "157   PIK3CA            0.494737      0.484211        0.010526\n",
              "166     PTEN            0.789474      0.463158        0.568421\n",
              "116     KRAS            0.326316      0.326316        0.000000\n",
              "55    CTNNB1            0.305263      0.305263        0.000000\n",
              "158   PIK3R1            0.389474      0.231579        0.189474\n",
              "134    MUC16            0.189474      0.178947        0.052632\n",
              "149    OBSCN            0.157895      0.157895        0.042105\n",
              "98     HUWE1            0.157895      0.157895        0.031579\n",
              "96     HERC2            0.178947      0.157895        0.063158\n",
              "87     FBXW7            0.189474      0.157895        0.042105\n",
              "205     TP53            0.221053      0.157895        0.073684\n",
              "223    ZFHX4            0.200000      0.147368        0.094737\n",
              "65    DNAH10            0.147368      0.147368        0.031579\n",
              "194    SYNE1            0.168421      0.147368        0.063158\n",
              "136    MUC5B            0.147368      0.136842        0.031579"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "361gzch-bgkF"
      },
      "source": [
        "In the table above, you'll find our favorite gene, TP53, has a missense mutation in 15% of our population. That's the gene we're going to spend the rest of our time on today, but you could also explore your favorite cancer gene."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZUvnCRSoifSI"
      },
      "source": [
        "# Part 2. TP53 and tumor histology\n",
        "\n",
        "TP53 is a popular cancer gene and is involved in DNA damage response (more [here](https://en.wikipedia.org/wiki/P53)). Today we'll look at the association between TP53 mutation status and the clinical characterization of tumors. To do this, let's join genomic data with clinical data. I'll select out only the histologic_type column, which is what tells us if the tumor is endometroid or serous. This clinical feature will be joined to the mutation data for TP53. A handy function in our package does all that for you. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 410
        },
        "id": "Ed2bUHRwi80r",
        "outputId": "31ca0fb2-1b60-487d-dd56-1539f3dfeb40"
      },
      "source": [
        "tp53_clinical_df = endo.join_metadata_to_mutations(metadata_df_name='clinical', mutations_genes='TP53', metadata_cols=['Histologic_type'], tissue_type=\"tumor\")\n",
        "tp53_clinical_df.head(10)\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "cptac warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 74 samples for the TP53 gene (/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py, line 1)\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Name</th>\n",
              "      <th>Histologic_type</th>\n",
              "      <th>TP53_Mutation</th>\n",
              "      <th>TP53_Location</th>\n",
              "      <th>TP53_Mutation_Status</th>\n",
              "      <th>Sample_Status</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Patient_ID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>C3L-00006</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Missense_Mutation]</td>\n",
              "      <td>[p.R248W]</td>\n",
              "      <td>Single_mutation</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00008</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00032</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00090</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00098</th>\n",
              "      <td>Serous</td>\n",
              "      <td>[Missense_Mutation]</td>\n",
              "      <td>[p.S241C]</td>\n",
              "      <td>Single_mutation</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00136</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00137</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Missense_Mutation]</td>\n",
              "      <td>[p.S127Y]</td>\n",
              "      <td>Single_mutation</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00139</th>\n",
              "      <td>Serous</td>\n",
              "      <td>[Nonsense_Mutation]</td>\n",
              "      <td>[p.R196*]</td>\n",
              "      <td>Single_mutation</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00143</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00145</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "Name       Histologic_type  ... Sample_Status\n",
              "Patient_ID                  ...              \n",
              "C3L-00006     Endometrioid  ...         Tumor\n",
              "C3L-00008     Endometrioid  ...         Tumor\n",
              "C3L-00032     Endometrioid  ...         Tumor\n",
              "C3L-00090     Endometrioid  ...         Tumor\n",
              "C3L-00098           Serous  ...         Tumor\n",
              "C3L-00136     Endometrioid  ...         Tumor\n",
              "C3L-00137     Endometrioid  ...         Tumor\n",
              "C3L-00139           Serous  ...         Tumor\n",
              "C3L-00143     Endometrioid  ...         Tumor\n",
              "C3L-00145     Endometrioid  ...         Tumor\n",
              "\n",
              "[10 rows x 5 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "taKw_9eXjz3o"
      },
      "source": [
        "I'm really interested in the overlap between serous tumors and TP53 mutation, and want to test whether there is a meaningful association between the two. This can be done with a Fisher's exact test. \n",
        "\n",
        "I am going to create a new column in my dataframe encoding the overlap. It can be thought of as the 2x2 matrix below"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "INst7vZh01Yg"
      },
      "source": [
        ". | Serous | Endometroid\n",
        " --- | --- | ----\n",
        "Mutant | cat-1 | cat-3\n",
        "Wildtype | cat-2 | cat-4"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "piobo_ejjzIX"
      },
      "source": [
        "def mut_hist_overlap(row):\n",
        "  if row['Histologic_type'] == 'Serous':\n",
        "    if row['TP53_Mutation_Status'] == \"Wildtype_Tumor\":\n",
        "      return 2\n",
        "    else:\n",
        "      return 1\n",
        "  else:\n",
        "    #not serous\n",
        "    if row['TP53_Mutation_Status'] == \"Wildtype_Tumor\":\n",
        "      return 4\n",
        "    else:\n",
        "      return 3\n",
        "  return 0 #failure\n",
        "\n",
        "#now apply the above function to make a new column\n",
        "tp53_clinical_df['overlap_category'] = tp53_clinical_df.apply(mut_hist_overlap, axis=1)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3kYyKRB1wow8"
      },
      "source": [
        "We can print off the dataframe to inspect the new row and make sure it looks like what we want. Notice the new column called 'overlap_category' which classifies each tumor into one of our four categories."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 373
        },
        "id": "wHR2avUFwjIZ",
        "outputId": "1e143d77-4039-41fb-cc3f-d2ae548d5bcb"
      },
      "source": [
        "tp53_clinical_df.head(10)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Name</th>\n",
              "      <th>Histologic_type</th>\n",
              "      <th>TP53_Mutation</th>\n",
              "      <th>TP53_Location</th>\n",
              "      <th>TP53_Mutation_Status</th>\n",
              "      <th>Sample_Status</th>\n",
              "      <th>overlap_category</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Patient_ID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>C3L-00006</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Missense_Mutation]</td>\n",
              "      <td>[p.R248W]</td>\n",
              "      <td>Single_mutation</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00008</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00032</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00090</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00098</th>\n",
              "      <td>Serous</td>\n",
              "      <td>[Missense_Mutation]</td>\n",
              "      <td>[p.S241C]</td>\n",
              "      <td>Single_mutation</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00136</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00137</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Missense_Mutation]</td>\n",
              "      <td>[p.S127Y]</td>\n",
              "      <td>Single_mutation</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00139</th>\n",
              "      <td>Serous</td>\n",
              "      <td>[Nonsense_Mutation]</td>\n",
              "      <td>[p.R196*]</td>\n",
              "      <td>Single_mutation</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00143</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00145</th>\n",
              "      <td>Endometrioid</td>\n",
              "      <td>[Wildtype_Tumor]</td>\n",
              "      <td>[No_mutation]</td>\n",
              "      <td>Wildtype_Tumor</td>\n",
              "      <td>Tumor</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "Name       Histologic_type        TP53_Mutation  ... Sample_Status overlap_category\n",
              "Patient_ID                                       ...                               \n",
              "C3L-00006     Endometrioid  [Missense_Mutation]  ...         Tumor                3\n",
              "C3L-00008     Endometrioid     [Wildtype_Tumor]  ...         Tumor                4\n",
              "C3L-00032     Endometrioid     [Wildtype_Tumor]  ...         Tumor                4\n",
              "C3L-00090     Endometrioid     [Wildtype_Tumor]  ...         Tumor                4\n",
              "C3L-00098           Serous  [Missense_Mutation]  ...         Tumor                1\n",
              "C3L-00136     Endometrioid     [Wildtype_Tumor]  ...         Tumor                4\n",
              "C3L-00137     Endometrioid  [Missense_Mutation]  ...         Tumor                3\n",
              "C3L-00139           Serous  [Nonsense_Mutation]  ...         Tumor                1\n",
              "C3L-00143     Endometrioid     [Wildtype_Tumor]  ...         Tumor                4\n",
              "C3L-00145     Endometrioid     [Wildtype_Tumor]  ...         Tumor                4\n",
              "\n",
              "[10 rows x 6 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "G-X4r9Yaw4SH"
      },
      "source": [
        "Now we will test to see whether histology and TP53 mutation are significantly associated using a Fisher's test. The Fisher's test is part of the stats package from SciPy, which we import below."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Uu1x5H5l4-XW"
      },
      "source": [
        "The Fisher's test tells us whether the imbalance we see in the 2x2 contingency table is likely to be random. Let's print out the contency table here with the numbers filled in.\n",
        "\n",
        ". | Serous | Endometroid\n",
        "--- | --- | ----\n",
        "Mutant | 12 | 9\n",
        "Wildtype | 0 | 74\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VfG_74Zn14I3",
        "outputId": "1bf7f9ef-6970-410e-bcfb-53c446fffbae"
      },
      "source": [
        "import scipy.stats as stats\n",
        "#give it two arrays. From the 2x2 matrix above we would give it the row-wise \n",
        "#values in our 2x2 matrix. For us that would be the categories as:\n",
        "#[[1,3], [2,4]]\n",
        "counts = tp53_clinical_df.value_counts('overlap_category')\n",
        "one = counts.get(1,0)#zero is the default value if this key doesn't exist in the counts series\n",
        "two = counts.get(2,0)\n",
        "three = counts.get(3,0)\n",
        "four = counts.get(4,0)\n",
        "oddsratio, pvalue = stats.fisher_exact([[one, three], [two, four]])\n",
        "print(pvalue)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "5.377570934016993e-10\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "K0yuoctIjsGr"
      },
      "source": [
        "So our p-value returned from the Fisher's test is very small and indicates that Histology and TP53 mutation are associated variables.\n",
        "\n",
        "There are a lot of serous tumors that are TP53-mut. And no serous tumors that are TP53-wt.  So it appears that TP53 mutation status has a strong influence on the histologic type of endometrial cancer. You can read more about this in our manuscript ([pubmed](https://pubmed.ncbi.nlm.nih.gov/32059776/), [Cell](https://www.sciencedirect.com/science/article/pii/S0092867420301070))."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EdQ480iQ6ndz"
      },
      "source": [
        "# Part 3. TP53 and molecular phenotype\n",
        "In this final part, we'll explore what happens to proteome abundance when TP53 is mutated. My specific question is what happens to proteins that are known to interact with TP53? \n",
        "\n",
        "In order to do this, I will get a list of interacting proteins. The cptac package has a few options for this, but I am going to choose WikiPathways. As you can see in the code below, WikiPathways annotates TP53 as belonging to 48 different pathway definitions. Instead of grabbing all of those, I am going to get only those that are annotated as part of the 'DNA Damage Response' or 'G1 to S cell cycle control' pathway. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Pq4dPoyHuMwk",
        "outputId": "ccbeff33-0086-4f7d-8b9d-85b739dbe127"
      },
      "source": [
        "pathways = utils.get_pathways_with_proteins(proteins ='TP53', database = 'wikipathways')\n",
        "print (pathways)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "      id                                            pathway\n",
            "0   TP53                             MAPK Signaling Pathway\n",
            "1   TP53                                DNA Damage Response\n",
            "2   TP53             Wnt Signaling Pathway and Pluripotency\n",
            "3   TP53      Pathways Affected in Adenoid Cystic Carcinoma\n",
            "4   TP53                      Retinoblastoma Gene in Cancer\n",
            "5   TP53  Mammary gland development pathway - Involution...\n",
            "6   TP53  miRNA regulation of p53 pathway in prostate ca...\n",
            "7   TP53            miRNA Regulation of DNA Damage Response\n",
            "8   TP53  Ultraconserved region 339 modulation of tumor ...\n",
            "9   TP53                           Gastric Cancer Network 2\n",
            "10  TP53                                       TP53 Network\n",
            "11  TP53  Chromosomal and microsatellite instability in ...\n",
            "12  TP53                     Circadian rhythm related genes\n",
            "13  TP53                              Breast cancer pathway\n",
            "14  TP53                         RAC1/PAK1/p38/MMP2 Pathway\n",
            "15  TP53                                          Apoptosis\n",
            "16  TP53           Hepatitis C and Hepatocellular Carcinoma\n",
            "17  TP53  The effect of progerin on the involved genes i...\n",
            "18  TP53                         G1 to S cell cycle control\n",
            "19  TP53                          Fluoropyrimidine Activity\n",
            "20  TP53                         PI3K-Akt Signaling Pathway\n",
            "21  TP53                                           Melanoma\n",
            "22  TP53                  Pancreatic adenocarcinoma pathway\n",
            "23  TP53                                 Spinal Cord Injury\n",
            "24  TP53                 Apoptosis Modulation and Signaling\n",
            "25  TP53                             Metastatic brain tumor\n",
            "26  TP53                         TGF-beta Signaling Pathway\n",
            "27  TP53  LncRNA-mediated mechanisms of therapeutic resi...\n",
            "28  TP53              Head and Neck Squamous Cell Carcinoma\n",
            "29  TP53        DNA IR-damage and cellular response via ATR\n",
            "30  TP53                         Non-small cell lung cancer\n",
            "31  TP53  MicroRNA network associated with chronic lymph...\n",
            "32  TP53                                        Ferroptosis\n",
            "33  TP53                 Signaling Pathways in Glioblastoma\n",
            "34  TP53                                 Alzheimers Disease\n",
            "35  TP53                                     Bladder Cancer\n",
            "36  TP53  Photodynamic therapy-induced HIF-1 survival si...\n",
            "37  TP53                            IL-18 signaling pathway\n",
            "38  TP53  Epithelial to mesenchymal transition in colore...\n",
            "39  TP53                                 Endometrial cancer\n",
            "40  TP53             miRNAs involved in DNA damage response\n",
            "41  TP53  DNA IR-Double Strand Breaks (DSBs) and cellula...\n",
            "42  TP53                 Senescence and Autophagy in Cancer\n",
            "43  TP53                                 Copper homeostasis\n",
            "44  TP53                Amyotrophic lateral sclerosis (ALS)\n",
            "45  TP53                              ATM Signaling Pathway\n",
            "46  TP53                     Netrin-UNC5B signaling Pathway\n",
            "47  TP53                             Small cell lung cancer\n",
            "48  TP53           DNA Damage Response (only ATM dependent)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "g0zK3O0Kt053",
        "outputId": "33db7f6b-8caf-464e-cda7-b4e29a518f5a"
      },
      "source": [
        "\n",
        "trimmed_pathways = ['DNA Damage Response', 'G1 to S cell cycle control']\n",
        "trans_genes = utils.get_proteins_in_pathways(pathways = trimmed_pathways, database = 'wikipathways')\n",
        "trans_genes_list = trans_genes['member'].tolist()\n",
        "#because I merged lists from multiple pathways, my list variable has duplicates\n",
        "#this quick list->set->list conversion will remove those\n",
        "trans_genes_list = list(set(trans_genes_list))\n",
        "print(trans_genes_list)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "['E2F3', 'CDK7', 'GADD45A', 'H2AFX', 'CDKN2B', 'CREB3L1', 'CREB1', 'MNAT1', 'MCM2', 'CREBL1', 'CCND1', 'MDM2', 'CCNB1', 'WEE1', 'CHEK2', 'NBN', 'CCNH', 'ORC1L', 'CCNA1', 'CDK6', 'CDKN1B', 'ORC2L', 'DDB2', 'ORC5L', 'TP53AIP1', 'TFDP2', 'CREB3L3', 'MCM5', 'ORC3L', 'CASP8', 'RAD1', 'BAX', 'ATRIP', 'SMC1A', 'ABL1', 'POLA2', 'TFDP1', 'CCNB2', 'PRKDC', 'MCM4', 'MYC', 'CDK4', 'RPA2', 'ORC4L', 'CDKN1A', 'CDC25C', 'CDKN2A', 'MCM3', 'E2F2', 'CDKN2D', 'CCNG2', 'CYCS', 'CASP9', 'PML', 'ATM', 'FANCD2', 'CCND2', 'MCM7', 'TLK1', 'BRCA1', 'CASP3', 'RPA1', 'RAD52', 'RRM2B', 'TNFRSF10B', 'CHEK1', 'SESN1', 'BID', 'FAS', 'POLE', 'RB1', 'PCNA', 'ORC6L', 'PIDD', 'CDK2', 'CCND3', 'PMAIP1', 'CDC45L', 'SFN', 'CDKN2C', 'MYT1', 'CREB3', 'RAD9A', 'CDC25A', 'GADD45B', 'CDK5', 'CREB3L4', 'CCNE1', 'POLE2', 'RPA3', 'CDKN1C', 'MCM6', 'TP53', 'CDK1', 'MRE11', 'CCNE2', 'ATR', 'HUS1', 'RAD50', 'PRIM2A', 'CDC2', 'BBC3', 'RAD17', 'GADD45G', 'CCNB3', 'PRIM1', 'TLK2', 'RAD51', 'E2F1', 'APAF1']\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_Ni2e_ugwcuo"
      },
      "source": [
        "After finding this list of genes, we want to check to see if their protein (or phosphorylation) abundance is altered in TP53 mutant tumors. In preparation for this, we want to create a dataframe that contains protein abundance and TP53 mutation status. This is done with another convenient join function"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 485
        },
        "id": "5fbNOEWVy9Qm",
        "outputId": "6467ebba-c87a-4020-cae2-0d0e7a1cf14d"
      },
      "source": [
        "mut_and_proteomics = endo.join_omics_to_mutations(omics_df_name='proteomics', mutations_genes='TP53', omics_genes=trans_genes_list, tissue_type = \"tumor\")\n",
        "\n",
        "#creating a binary value column with either wildtype or mutation\n",
        "#the join function natively returns quite a lot of detail about the mutation\n",
        "#which I don't really want for the t-tests.\n",
        "def simplify_status(row):\n",
        "  if row['TP53_Mutation_Status'] == \"Wildtype_Tumor\":\n",
        "    return \"Wildtype\"\n",
        "  else:\n",
        "    return \"Mutation\"\n",
        "\n",
        "\n",
        "mut_and_proteomics['Simple_mut_status'] = mut_and_proteomics.apply(simplify_status, axis=1)\n",
        "mut_and_proteomics.drop(columns = ['TP53_Mutation', 'TP53_Location', 'Sample_Status', 'TP53_Mutation_Status'], inplace=True)\n",
        "mut_and_proteomics"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "cptac warning: The following columns were not found in the proteomics dataframe, so they were inserted into joined table, but filled with NaN: BRCA1, CCNA1, CCNB3, CCND3, CCNE2, CCNG2, CDC2, CDC25A, CDC25C, CDC45L, CDKN2B, CDKN2D, CREB3, CREB3L3, CREBL1, E2F1, E2F2, GADD45A, GADD45B, GADD45G, MDM2, MYC, MYT1, ORC1L, ORC2L, ORC3L, ORC4L, ORC5L, ORC6L, PIDD, PRIM2A, RAD51, RAD52, TP53AIP1 (/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py, line 1)\n",
            "cptac warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 74 samples for the TP53 gene (/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py, line 1)\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Name</th>\n",
              "      <th>E2F3_proteomics</th>\n",
              "      <th>CDK7_proteomics</th>\n",
              "      <th>GADD45A_proteomics</th>\n",
              "      <th>H2AFX_proteomics</th>\n",
              "      <th>CDKN2B_proteomics</th>\n",
              "      <th>CREB3L1_proteomics</th>\n",
              "      <th>CREB1_proteomics</th>\n",
              "      <th>MNAT1_proteomics</th>\n",
              "      <th>MCM2_proteomics</th>\n",
              "      <th>CREBL1_proteomics</th>\n",
              "      <th>CCND1_proteomics</th>\n",
              "      <th>MDM2_proteomics</th>\n",
              "      <th>CCNB1_proteomics</th>\n",
              "      <th>WEE1_proteomics</th>\n",
              "      <th>CHEK2_proteomics</th>\n",
              "      <th>NBN_proteomics</th>\n",
              "      <th>CCNH_proteomics</th>\n",
              "      <th>ORC1L_proteomics</th>\n",
              "      <th>CCNA1_proteomics</th>\n",
              "      <th>CDK6_proteomics</th>\n",
              "      <th>CDKN1B_proteomics</th>\n",
              "      <th>ORC2L_proteomics</th>\n",
              "      <th>DDB2_proteomics</th>\n",
              "      <th>ORC5L_proteomics</th>\n",
              "      <th>TP53AIP1_proteomics</th>\n",
              "      <th>TFDP2_proteomics</th>\n",
              "      <th>CREB3L3_proteomics</th>\n",
              "      <th>MCM5_proteomics</th>\n",
              "      <th>ORC3L_proteomics</th>\n",
              "      <th>CASP8_proteomics</th>\n",
              "      <th>RAD1_proteomics</th>\n",
              "      <th>BAX_proteomics</th>\n",
              "      <th>ATRIP_proteomics</th>\n",
              "      <th>SMC1A_proteomics</th>\n",
              "      <th>ABL1_proteomics</th>\n",
              "      <th>POLA2_proteomics</th>\n",
              "      <th>TFDP1_proteomics</th>\n",
              "      <th>CCNB2_proteomics</th>\n",
              "      <th>PRKDC_proteomics</th>\n",
              "      <th>MCM4_proteomics</th>\n",
              "      <th>...</th>\n",
              "      <th>PCNA_proteomics</th>\n",
              "      <th>ORC6L_proteomics</th>\n",
              "      <th>PIDD_proteomics</th>\n",
              "      <th>CDK2_proteomics</th>\n",
              "      <th>CCND3_proteomics</th>\n",
              "      <th>PMAIP1_proteomics</th>\n",
              "      <th>CDC45L_proteomics</th>\n",
              "      <th>SFN_proteomics</th>\n",
              "      <th>CDKN2C_proteomics</th>\n",
              "      <th>MYT1_proteomics</th>\n",
              "      <th>CREB3_proteomics</th>\n",
              "      <th>RAD9A_proteomics</th>\n",
              "      <th>CDC25A_proteomics</th>\n",
              "      <th>GADD45B_proteomics</th>\n",
              "      <th>CDK5_proteomics</th>\n",
              "      <th>CREB3L4_proteomics</th>\n",
              "      <th>CCNE1_proteomics</th>\n",
              "      <th>POLE2_proteomics</th>\n",
              "      <th>RPA3_proteomics</th>\n",
              "      <th>CDKN1C_proteomics</th>\n",
              "      <th>MCM6_proteomics</th>\n",
              "      <th>TP53_proteomics</th>\n",
              "      <th>CDK1_proteomics</th>\n",
              "      <th>MRE11_proteomics</th>\n",
              "      <th>CCNE2_proteomics</th>\n",
              "      <th>ATR_proteomics</th>\n",
              "      <th>HUS1_proteomics</th>\n",
              "      <th>RAD50_proteomics</th>\n",
              "      <th>PRIM2A_proteomics</th>\n",
              "      <th>CDC2_proteomics</th>\n",
              "      <th>BBC3_proteomics</th>\n",
              "      <th>RAD17_proteomics</th>\n",
              "      <th>GADD45G_proteomics</th>\n",
              "      <th>CCNB3_proteomics</th>\n",
              "      <th>PRIM1_proteomics</th>\n",
              "      <th>TLK2_proteomics</th>\n",
              "      <th>RAD51_proteomics</th>\n",
              "      <th>E2F1_proteomics</th>\n",
              "      <th>APAF1_proteomics</th>\n",
              "      <th>Simple_mut_status</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Patient_ID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>C3L-00006</th>\n",
              "      <td>NaN</td>\n",
              "      <td>0.20500</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.05970</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.5720</td>\n",
              "      <td>0.1650</td>\n",
              "      <td>0.6150</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.2170</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.2900</td>\n",
              "      <td>0.34400</td>\n",
              "      <td>1.450</td>\n",
              "      <td>-1.0600</td>\n",
              "      <td>-0.0552</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.1770</td>\n",
              "      <td>-0.9950</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.4430</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.2780</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.6890</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.105</td>\n",
              "      <td>0.2290</td>\n",
              "      <td>-0.27600</td>\n",
              "      <td>0.3090</td>\n",
              "      <td>-0.2500</td>\n",
              "      <td>-0.2960</td>\n",
              "      <td>0.5540</td>\n",
              "      <td>0.1830</td>\n",
              "      <td>0.921</td>\n",
              "      <td>0.4420</td>\n",
              "      <td>0.759</td>\n",
              "      <td>...</td>\n",
              "      <td>0.800</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.5150</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.22300</td>\n",
              "      <td>-1.2100</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.731000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.5190</td>\n",
              "      <td>-0.00828</td>\n",
              "      <td>0.0275</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0235</td>\n",
              "      <td>-0.3810</td>\n",
              "      <td>0.859</td>\n",
              "      <td>0.2950</td>\n",
              "      <td>1.100</td>\n",
              "      <td>-1.36000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.17800</td>\n",
              "      <td>0.3740</td>\n",
              "      <td>-1.9100</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.8640</td>\n",
              "      <td>0.1770</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.42800</td>\n",
              "      <td>-0.4940</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0693</td>\n",
              "      <td>Mutation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00008</th>\n",
              "      <td>0.2050</td>\n",
              "      <td>0.27600</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.00904</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.6830</td>\n",
              "      <td>-0.0190</td>\n",
              "      <td>0.9100</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2.8900</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.0800</td>\n",
              "      <td>0.25700</td>\n",
              "      <td>0.953</td>\n",
              "      <td>-0.1440</td>\n",
              "      <td>0.3020</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-1.1700</td>\n",
              "      <td>0.6010</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0791</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0394</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.7750</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.325</td>\n",
              "      <td>0.2780</td>\n",
              "      <td>0.65600</td>\n",
              "      <td>-0.7920</td>\n",
              "      <td>-0.4150</td>\n",
              "      <td>-0.0897</td>\n",
              "      <td>0.8960</td>\n",
              "      <td>0.7870</td>\n",
              "      <td>0.392</td>\n",
              "      <td>0.2130</td>\n",
              "      <td>1.000</td>\n",
              "      <td>...</td>\n",
              "      <td>1.010</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.1600</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.10600</td>\n",
              "      <td>-0.9920</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.002390</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0172</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.1600</td>\n",
              "      <td>0.80000</td>\n",
              "      <td>0.1920</td>\n",
              "      <td>-0.2960</td>\n",
              "      <td>1.070</td>\n",
              "      <td>0.2770</td>\n",
              "      <td>1.140</td>\n",
              "      <td>-0.27500</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.00558</td>\n",
              "      <td>0.3990</td>\n",
              "      <td>-0.2270</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.5430</td>\n",
              "      <td>-0.1900</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.88200</td>\n",
              "      <td>-0.2230</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.2200</td>\n",
              "      <td>Wildtype</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00032</th>\n",
              "      <td>0.2350</td>\n",
              "      <td>-0.03490</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.28700</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0471</td>\n",
              "      <td>-0.1400</td>\n",
              "      <td>-0.0715</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.5510</td>\n",
              "      <td>-0.03000</td>\n",
              "      <td>0.659</td>\n",
              "      <td>-0.0141</td>\n",
              "      <td>-0.0643</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.2640</td>\n",
              "      <td>-0.0728</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.6030</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.4710</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.2580</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.364</td>\n",
              "      <td>0.1130</td>\n",
              "      <td>-0.00859</td>\n",
              "      <td>0.0478</td>\n",
              "      <td>-0.3570</td>\n",
              "      <td>-0.1770</td>\n",
              "      <td>-0.1750</td>\n",
              "      <td>0.1830</td>\n",
              "      <td>-0.477</td>\n",
              "      <td>0.0277</td>\n",
              "      <td>-0.440</td>\n",
              "      <td>...</td>\n",
              "      <td>-0.285</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.2040</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.00883</td>\n",
              "      <td>-0.1350</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.155000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.2740</td>\n",
              "      <td>0.54700</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.09740</td>\n",
              "      <td>-0.6030</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.210</td>\n",
              "      <td>-0.8710</td>\n",
              "      <td>0.220</td>\n",
              "      <td>-0.00493</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.30400</td>\n",
              "      <td>-0.2310</td>\n",
              "      <td>0.0503</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.2600</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.12100</td>\n",
              "      <td>-0.4240</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0317</td>\n",
              "      <td>Wildtype</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00090</th>\n",
              "      <td>0.0346</td>\n",
              "      <td>0.15400</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.13200</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.126</td>\n",
              "      <td>-0.5510</td>\n",
              "      <td>0.1480</td>\n",
              "      <td>-0.5630</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.4990</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0836</td>\n",
              "      <td>-0.64300</td>\n",
              "      <td>-1.620</td>\n",
              "      <td>-0.0684</td>\n",
              "      <td>0.3660</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0780</td>\n",
              "      <td>-0.5200</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.4910</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.3360</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.3500</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.130</td>\n",
              "      <td>-0.2230</td>\n",
              "      <td>0.31300</td>\n",
              "      <td>-0.4730</td>\n",
              "      <td>-0.5300</td>\n",
              "      <td>-0.5130</td>\n",
              "      <td>0.0747</td>\n",
              "      <td>-0.3420</td>\n",
              "      <td>-0.146</td>\n",
              "      <td>-0.0491</td>\n",
              "      <td>-0.515</td>\n",
              "      <td>...</td>\n",
              "      <td>0.296</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0604</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.46000</td>\n",
              "      <td>-1.1900</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.140000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0195</td>\n",
              "      <td>-0.67200</td>\n",
              "      <td>0.1900</td>\n",
              "      <td>-0.44700</td>\n",
              "      <td>-0.1100</td>\n",
              "      <td>-0.6350</td>\n",
              "      <td>-0.350</td>\n",
              "      <td>-0.3430</td>\n",
              "      <td>0.505</td>\n",
              "      <td>-0.20800</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.35100</td>\n",
              "      <td>-0.1830</td>\n",
              "      <td>-0.1730</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.1470</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.01540</td>\n",
              "      <td>0.0139</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.1370</td>\n",
              "      <td>Wildtype</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3L-00098</th>\n",
              "      <td>0.6750</td>\n",
              "      <td>0.10600</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-1.48000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.3150</td>\n",
              "      <td>0.1120</td>\n",
              "      <td>1.4100</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-1.5200</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.7400</td>\n",
              "      <td>0.10400</td>\n",
              "      <td>1.640</td>\n",
              "      <td>0.0777</td>\n",
              "      <td>0.1620</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-1.4200</td>\n",
              "      <td>-0.8800</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-1.9500</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.7790</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.5900</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.420</td>\n",
              "      <td>0.4020</td>\n",
              "      <td>-0.81900</td>\n",
              "      <td>-0.1620</td>\n",
              "      <td>-0.4090</td>\n",
              "      <td>-0.5260</td>\n",
              "      <td>0.4630</td>\n",
              "      <td>0.7400</td>\n",
              "      <td>1.230</td>\n",
              "      <td>0.4350</td>\n",
              "      <td>1.660</td>\n",
              "      <td>...</td>\n",
              "      <td>1.220</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.3920</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.28000</td>\n",
              "      <td>1.3200</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.146000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.2970</td>\n",
              "      <td>-0.02350</td>\n",
              "      <td>0.7720</td>\n",
              "      <td>-0.00505</td>\n",
              "      <td>-0.1700</td>\n",
              "      <td>-1.5800</td>\n",
              "      <td>1.740</td>\n",
              "      <td>3.0100</td>\n",
              "      <td>1.680</td>\n",
              "      <td>-0.48300</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.01930</td>\n",
              "      <td>0.3020</td>\n",
              "      <td>-0.4200</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.8320</td>\n",
              "      <td>-0.7970</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.34000</td>\n",
              "      <td>0.2270</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.5230</td>\n",
              "      <td>Mutation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3N-01520</th>\n",
              "      <td>0.4940</td>\n",
              "      <td>0.02460</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.21000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.506</td>\n",
              "      <td>-0.2360</td>\n",
              "      <td>0.1330</td>\n",
              "      <td>0.4810</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.2370</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.8610</td>\n",
              "      <td>-0.00265</td>\n",
              "      <td>0.885</td>\n",
              "      <td>-1.0600</td>\n",
              "      <td>-0.1450</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.5160</td>\n",
              "      <td>-1.6800</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.4740</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.3880</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.6760</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.265</td>\n",
              "      <td>-0.1630</td>\n",
              "      <td>0.46500</td>\n",
              "      <td>0.6370</td>\n",
              "      <td>-0.3280</td>\n",
              "      <td>-0.1810</td>\n",
              "      <td>0.1480</td>\n",
              "      <td>0.1310</td>\n",
              "      <td>-0.102</td>\n",
              "      <td>0.1710</td>\n",
              "      <td>0.728</td>\n",
              "      <td>...</td>\n",
              "      <td>0.663</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.2800</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.01120</td>\n",
              "      <td>-1.1400</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.050300</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.0600</td>\n",
              "      <td>-0.48500</td>\n",
              "      <td>0.4610</td>\n",
              "      <td>0.73400</td>\n",
              "      <td>0.1570</td>\n",
              "      <td>-0.0661</td>\n",
              "      <td>0.686</td>\n",
              "      <td>-0.2570</td>\n",
              "      <td>0.918</td>\n",
              "      <td>-1.39000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.56000</td>\n",
              "      <td>0.1040</td>\n",
              "      <td>-1.5300</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0847</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.32500</td>\n",
              "      <td>0.0987</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0627</td>\n",
              "      <td>Wildtype</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3N-01521</th>\n",
              "      <td>0.1600</td>\n",
              "      <td>0.00676</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.53300</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0321</td>\n",
              "      <td>-0.2790</td>\n",
              "      <td>0.2310</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.1190</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.5040</td>\n",
              "      <td>0.70100</td>\n",
              "      <td>-0.419</td>\n",
              "      <td>-0.0867</td>\n",
              "      <td>-0.2150</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.2460</td>\n",
              "      <td>-0.3330</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.3420</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.5470</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0119</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.182</td>\n",
              "      <td>0.1210</td>\n",
              "      <td>0.31600</td>\n",
              "      <td>-0.2360</td>\n",
              "      <td>0.1290</td>\n",
              "      <td>-0.6250</td>\n",
              "      <td>0.2740</td>\n",
              "      <td>-0.2280</td>\n",
              "      <td>0.627</td>\n",
              "      <td>0.2830</td>\n",
              "      <td>0.300</td>\n",
              "      <td>...</td>\n",
              "      <td>0.822</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0819</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.33800</td>\n",
              "      <td>-0.9860</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.002550</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.7200</td>\n",
              "      <td>0.43800</td>\n",
              "      <td>-0.1160</td>\n",
              "      <td>-0.23900</td>\n",
              "      <td>-0.1860</td>\n",
              "      <td>-0.3540</td>\n",
              "      <td>0.338</td>\n",
              "      <td>1.2000</td>\n",
              "      <td>0.426</td>\n",
              "      <td>-0.34300</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.12900</td>\n",
              "      <td>0.0126</td>\n",
              "      <td>-0.4900</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0734</td>\n",
              "      <td>-0.2140</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.29900</td>\n",
              "      <td>-0.1420</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.4160</td>\n",
              "      <td>Mutation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3N-01537</th>\n",
              "      <td>-0.0847</td>\n",
              "      <td>0.39700</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.03710</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.323</td>\n",
              "      <td>-0.5040</td>\n",
              "      <td>0.1620</td>\n",
              "      <td>0.7840</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.2270</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.0600</td>\n",
              "      <td>-0.13800</td>\n",
              "      <td>-1.520</td>\n",
              "      <td>0.0449</td>\n",
              "      <td>0.3210</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.5430</td>\n",
              "      <td>0.2110</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.7700</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0809</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.4240</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.950</td>\n",
              "      <td>0.0514</td>\n",
              "      <td>0.36100</td>\n",
              "      <td>0.1630</td>\n",
              "      <td>-0.0761</td>\n",
              "      <td>-0.2450</td>\n",
              "      <td>0.3850</td>\n",
              "      <td>0.1550</td>\n",
              "      <td>0.976</td>\n",
              "      <td>0.1730</td>\n",
              "      <td>0.836</td>\n",
              "      <td>...</td>\n",
              "      <td>-0.069</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.5950</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.44300</td>\n",
              "      <td>0.4770</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.047300</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.2510</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.1100</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0305</td>\n",
              "      <td>-0.6910</td>\n",
              "      <td>0.923</td>\n",
              "      <td>-0.0333</td>\n",
              "      <td>0.371</td>\n",
              "      <td>0.01060</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.08500</td>\n",
              "      <td>0.3120</td>\n",
              "      <td>0.1010</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0875</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.31000</td>\n",
              "      <td>0.1010</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0308</td>\n",
              "      <td>Wildtype</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3N-01802</th>\n",
              "      <td>0.2760</td>\n",
              "      <td>0.06370</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.21300</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.8010</td>\n",
              "      <td>0.0962</td>\n",
              "      <td>0.1380</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0974</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.6100</td>\n",
              "      <td>-0.14700</td>\n",
              "      <td>-0.521</td>\n",
              "      <td>0.5270</td>\n",
              "      <td>0.6990</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0907</td>\n",
              "      <td>0.4940</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-1.3300</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.1070</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.1780</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.398</td>\n",
              "      <td>0.0000</td>\n",
              "      <td>0.49000</td>\n",
              "      <td>-0.9960</td>\n",
              "      <td>-0.3100</td>\n",
              "      <td>0.1190</td>\n",
              "      <td>-0.2960</td>\n",
              "      <td>0.3660</td>\n",
              "      <td>1.220</td>\n",
              "      <td>0.1560</td>\n",
              "      <td>-0.119</td>\n",
              "      <td>...</td>\n",
              "      <td>0.368</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0563</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.28800</td>\n",
              "      <td>-1.9600</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.266000</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0156</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2.8000</td>\n",
              "      <td>0.06870</td>\n",
              "      <td>-0.9780</td>\n",
              "      <td>0.4420</td>\n",
              "      <td>-0.401</td>\n",
              "      <td>0.2920</td>\n",
              "      <td>0.374</td>\n",
              "      <td>-0.10700</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.41000</td>\n",
              "      <td>-0.1900</td>\n",
              "      <td>0.1470</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2.5000</td>\n",
              "      <td>0.4890</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.00748</td>\n",
              "      <td>-0.2850</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.1330</td>\n",
              "      <td>Mutation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>C3N-01825</th>\n",
              "      <td>-0.1490</td>\n",
              "      <td>0.11700</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.07420</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0491</td>\n",
              "      <td>-0.0472</td>\n",
              "      <td>0.7970</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0786</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.7050</td>\n",
              "      <td>-0.44200</td>\n",
              "      <td>0.368</td>\n",
              "      <td>-0.0580</td>\n",
              "      <td>0.0864</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.2020</td>\n",
              "      <td>-0.0858</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.1790</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0389</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.6970</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.393</td>\n",
              "      <td>0.4110</td>\n",
              "      <td>0.14700</td>\n",
              "      <td>0.0440</td>\n",
              "      <td>-0.2250</td>\n",
              "      <td>0.1480</td>\n",
              "      <td>0.3930</td>\n",
              "      <td>-0.0683</td>\n",
              "      <td>0.891</td>\n",
              "      <td>0.1930</td>\n",
              "      <td>0.603</td>\n",
              "      <td>...</td>\n",
              "      <td>0.158</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.3580</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.0766</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.42900</td>\n",
              "      <td>-0.0261</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.000726</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.3550</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0685</td>\n",
              "      <td>-0.29200</td>\n",
              "      <td>-0.0941</td>\n",
              "      <td>-0.2010</td>\n",
              "      <td>0.806</td>\n",
              "      <td>1.5100</td>\n",
              "      <td>0.514</td>\n",
              "      <td>-0.02480</td>\n",
              "      <td>NaN</td>\n",
              "      <td>-0.12400</td>\n",
              "      <td>-0.2520</td>\n",
              "      <td>-0.0793</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.6600</td>\n",
              "      <td>-0.3430</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.15500</td>\n",
              "      <td>0.0653</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0411</td>\n",
              "      <td>Mutation</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>95 rows × 111 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "Name        E2F3_proteomics  ...  Simple_mut_status\n",
              "Patient_ID                   ...                   \n",
              "C3L-00006               NaN  ...           Mutation\n",
              "C3L-00008            0.2050  ...           Wildtype\n",
              "C3L-00032            0.2350  ...           Wildtype\n",
              "C3L-00090            0.0346  ...           Wildtype\n",
              "C3L-00098            0.6750  ...           Mutation\n",
              "...                     ...  ...                ...\n",
              "C3N-01520            0.4940  ...           Wildtype\n",
              "C3N-01521            0.1600  ...           Mutation\n",
              "C3N-01537           -0.0847  ...           Wildtype\n",
              "C3N-01802            0.2760  ...           Mutation\n",
              "C3N-01825           -0.1490  ...           Mutation\n",
              "\n",
              "[95 rows x 111 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MBuq87gMyzJv"
      },
      "source": [
        "We can use a T-test as a simple univariate test to see whether TP53 mutant tumors have a significantly different abundance of a protein compared to wildtype tumors. As we have shown above, the TP53 mutation has a strong association with histologic subtype. In this case, it is difficult to determine whether TP53 or cancer subtype is responsible for the observed change in abundance. This is complicated by the fact that many believe that TP53 mutation drives, or causes, the serous subtype. For this reason, although other more complex multi-variate statistics can be used, we'll be using a T-test for this tutorial."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 402
        },
        "id": "vad0ZC2FwbzP",
        "outputId": "fd73a489-4157-4c94-d052-1eb4e28da1b8"
      },
      "source": [
        "results = utils.wrap_ttest(df=mut_and_proteomics, label_column = 'Simple_mut_status'  )\n",
        "results"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Comparison</th>\n",
              "      <th>P_Value</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>TP53_proteomics</td>\n",
              "      <td>0.000030</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>FAS_proteomics</td>\n",
              "      <td>0.000061</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>CCNB1_proteomics</td>\n",
              "      <td>0.001113</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>MCM2_proteomics</td>\n",
              "      <td>0.013348</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>MCM7_proteomics</td>\n",
              "      <td>0.021519</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>DDB2_proteomics</td>\n",
              "      <td>0.022133</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>MCM5_proteomics</td>\n",
              "      <td>0.027433</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>PRKDC_proteomics</td>\n",
              "      <td>0.032158</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>MCM6_proteomics</td>\n",
              "      <td>0.033835</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>MCM3_proteomics</td>\n",
              "      <td>0.035256</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>CDKN2A_proteomics</td>\n",
              "      <td>0.037653</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>CDK1_proteomics</td>\n",
              "      <td>0.039515</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "           Comparison   P_Value\n",
              "0     TP53_proteomics  0.000030\n",
              "1      FAS_proteomics  0.000061\n",
              "2    CCNB1_proteomics  0.001113\n",
              "3     MCM2_proteomics  0.013348\n",
              "4     MCM7_proteomics  0.021519\n",
              "5     DDB2_proteomics  0.022133\n",
              "6     MCM5_proteomics  0.027433\n",
              "7    PRKDC_proteomics  0.032158\n",
              "8     MCM6_proteomics  0.033835\n",
              "9     MCM3_proteomics  0.035256\n",
              "10  CDKN2A_proteomics  0.037653\n",
              "11    CDK1_proteomics  0.039515"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "gV1WFUvAAUXJ"
      },
      "source": [
        "So I see a lot of the MCM complex proteins in there. Let's do a quick box plot to visualize these results and see whether our proteins are up or down.\n",
        "\n",
        "We make the plot below with a graphing library called Seaborn. The chart shows that MCM2 protein abundance is significantly up in tumors with a TP53 mutation compared to wildtype. This is presumably because TP53 mutated tumors are going through the cell cycle more frequently and thus replication machinery is more abundant in the tissue."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 319
        },
        "id": "mXweLRGwAijc",
        "outputId": "248c8368-2c7b-48a4-a964-cccc42b16b75"
      },
      "source": [
        "import seaborn as sns\n",
        "\n",
        "plot = sns.boxplot(data = mut_and_proteomics, x = 'Simple_mut_status', y = 'MCM2_proteomics', showfliers = False)\n",
        "plot = sns.stripplot(data = mut_and_proteomics, x = 'Simple_mut_status', y = 'MCM2_proteomics', jitter = True, dodge = True, color = '.3')\n",
        "plot.set_xlabel(\"\\nTP53 mut status\", fontsize=15)\n",
        "plot.set_ylabel(\"MCM2 proteomics\", fontsize=15)"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Text(0, 0.5, 'MCM2 proteomics')"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 12
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY8AAAEdCAYAAAD0NOuvAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nOzdd3yUdbb48c9JT0gjEFqoEooo1chakFVUbKys176uq4truyuW1VV/rt27ruhdV8RlsetWrl10dS2I2AsIiCAlECCAhJCEtEmdOb8/nkmYJJNkhkwyKef9euWVPM9853nOhDBnvl1UFWOMMSYYEeEOwBhjTNdjycMYY0zQLHkYY4wJmiUPY4wxQbPkYYwxJmiWPIwxxgQtKtwBdIS+ffvq8OHDwx2GMcZ0KStXrtynqun+HusRyWP48OGsWLEi3GEYY0yXIiLbm3vMmq2MMcYEzZKHMcaYoFnyMMYYEzRLHsYYY4JmycMYY0zQesRoK2NM17RmzRo+/PBDevfuzU9+8hN69+4d7pCMlyUPY0yntGrVKu6++27qto34/PPPeeyxx4iMjAxzZAYseRhjQmDBggVkZ2eH9Jp79+7Fd7+hXbt2cfXVVxMfH9/sc3bt2gVARkZGSGMJVmZmJnPnzg1rDO3NkocxJuw8Hg/FxcXU1NSQkJBAYmIiERFNu2Rbq3VUVFS0V4imEUsexpg2a+un7DvvvJPt253JzOXl5Zx11llkZWVx6623UlhYCMAJJ5zADTfc0OJ1rrvuOgDmz5/fpnhM6yx5GGPCqqCggNWrVzc49/7773Paaadx2WWXsW7dOn70ox8xefLkMEVo/LHkYYwJq/j4eGJiYqiurq4/l5yczN13382aNWsA+Pjjj3nooYcYNGhQuMI0jdg8D2NMWCUkJHDBBRcgIvXHRx99dH3iACgtLeXNN98MV4jGD6t5GGPC7pxzzuGYY45h586djBs3js2bNzcp41szMeHX6ZKHiDwDzAL2qurhfh4/HngdyPGeekVV7+24CE0wXC4Xr776Ktu3b2fKlCmccsop9Z8wjfE1aNCg+mapCRMmMGzYsPpO9OjoaE499dRwhmca6XTJA3gOeAz4awtlPlbVWR0TjmmLefPmsWrVKgC++OILSkpKOO+888IclensIiMjeeCBB3j//fcpKSnhxz/+MUOHDg13WMZHp0seqvqRiAwPdxym7YqLi+sTR50PP/zQkocJSK9evZg9e3a4wzDN6Kod5keLyBoReVtEDgt3MMa/uLi4JrOBbW0iY7qHrpg8vgGGqepEYAHwmr9CInKFiKwQkRX5+fkdGqBxxMbGcskll9TPFO7VqxcXX3xxmKMyxoRCp2u2ao2qlvj8/JaILBSRvqq6r1G5J4AnALKyshQTFqeffjpTp04lNzeXMWPGkJCQEO6QjDEh0OWSh4gMAPJUVUVkKk7tqSDMYZkW9O3bl759+4Y7DGNMCHW65CEi/wKOB/qKyE7gLiAaQFUXAecAV4tILVABXKC+S2+asCgqKiIvL4/MzEyiojrdn5UxzSopKWH37t2MHDmS6OjocIfTZXS6/+WqemErjz+GM5TXdBJLlizh2Wefxe1207dvX+67776wL4ltTCCWLl3KwoULqampISUlhbvvvpuRI0eGO6wuodMlD9O89tgzIViN90twu93k5ubW77uwb98+brrpJvr169fusfSEPRNM+6muruapp56ipqYGcIaWP//889x7r805DkRXHG1lwkRVqayspKqqitzcXMrLy3G73TRuNaytrQ1ThMYEzuVyUV5e3uCcjcwMnNU8upBwf8p+5ZVX2LZtG+AkiKKiIp5++mnuuusucnJy6stdfPHFnHnmmU2ev379etasWcMhhxzC1KlTbZkSE1apqamMHz+etWvX1p877rjjwhhR12LJwwRsw4YNDY5ramrYsmULd911Fy+88AK7d+/mqKOO4rTTTmvy3HfffZfHHjvQVTV79mwuu+yydo/ZmJbccsstvPjii2zbto0jjjiCn/zkJ+EOqcuw5GECduihh/LFF1/UH8fExJCZmUlKSgpXXXVVi8999dVXGxy/9dZb/PznPyc2NrZdYjUmEMnJyfYh5iBZn4cJ2E9+8hOSkpIQEQYOHMhNN91ESkpKQM9tvB+1iFizlelQ5eXlvPnmm7zwwgvk5eWFO5wuz2oeJmBRUVH1E/6C3SP67LPPZv78+fWd62eeeSYxMTHtEaYxTdTU1HDzzTeTm5sLwMsvv8wf//hHBg8eHObIui5LHqZDzJgxg6FDh9Z3mNt+1KYjrVy5sj5xAFRUVPDuu+8yZ86cMEbVtVnyMB0mMzOTzMzMcIdheqDIyMiAzpnAWZ+HMabbmzx5coMPLsnJybYzYRtZzcMY0+1FRUXxhz/8gc8//xyXy8UxxxxDampquMPq0ix5GGN6hNjYWI4//vhwh9FtWLOVMabLU1Vee+01du/eTV5eXoPOcdM+LHkYY7q8t99+m2eeeYaqqipcLhd33XWXrbHWzix5GGO6vC+//LLB8b59+8K+AnV3Z8nDGNPlDRw4sMFxZGQk/fv3D1M0PYMlD2NMl3f++efXb+IkIsyZM4fevXuHOaru7aCTh4j0FpFJImIr2xljwqp379786U9/YvDgwQwZMsRWx+0AASUPEblHRB7wOZ4B7ABWAltE5LB2is8YYwIWHR1tM8c7SKA1j4sA380c/gh8AhwLbAT+EOK4TJhs3LiRzz77DJfLFe5QjDGdWKCTBAcBWwFEZAgwEbhSVb8SkYeBZ9spPtOBFixYwHvvvQc4yzc88MADtuqoMcavQGsepUDdxg0zgCJV/cp7XAkkhDow07F27txZnzgASkpKeOWVV8IYkTGmMwu05rEcuFVEPMBNwOs+j40GbDpnF1deXt7kXFlZWRgiMcZ0BYHWPG4AqoDFwH7gdz6P/QL4KMRxmQ42atQoRowYUX8sIpx88slhjMgY05kFVPNQ1V04zVX+nILTdGW6sIiICO677z7+/e9/U1BQwHHHHcfEiRPDHZYxppMKKHl4O8nTVfUbPw9nAvlY01WXl5yczIUXXhjuMIwxXUCgfR5/ATYB/pLHz4AxgM3KMcZ0GgUFBbz55puUlZVx4oknMnbs2HCH1K0EmjyOAhY189gy4JLQhGOMMW1XVVXFLbfcwt69ewF47733eOCBByyBhFCgHeYJgLbweK8QxAKAiDwjIntF5LtmHhcReVREskXkWxGZEqp7G2O6hzVr1tQnDgCPx8PSpUvDGFH3E2jyWAs01xh+IbAuNOEA8BzQ0ubCpwGjvF9X4DSpGWNMvcTExIDOmYMXaPJ4APiZiLwoImeIyBTv9xdwksfvQxWQqn4EFLZQZDbwV3V8AaSKyMAWyhtjephx48Zx1FFH1R+np6cza9asMEbU/QQ6VPdVEbkEZw2rs3GasATYBfxcVV9rvxCbyKDhyK6d3nM/dGAMxphO7rbbbmP9+vWUlpYyadIkYmNtAfBQCrTDHFX9m4j8HWdkVR+gANioqi31hYSNiFyB06zF0KFDwxxN51VaWsqePXsYPnw40dHRLZZ1u924XC5UlaqqKvvPaDq9cePGhTuEbiuo/Ty8TUUbVPVT7/dwJI5dwBCf48Hecw2o6hOqmqWqWenp6R0WXFeydOlSfvnLX3LjjTdy+eWXk5OT02xZt9vNHXfcQV5eHnv37uWaa65h//79Qd9zy5YtfPXVV1RVVbUldGNMmDVb8xCR/wZeVNV8788tUVXtqI7rJcA1IrIY+BFQrKrt2mS1YMGCbrcfssfjITc3F4/HA0BhYSG33norAwYM8Fve5XKRl5dXf5yXl8e1114b1G5t+/bto7S0FHC2CR04cGCrtZ3OLjMzk7lz54Y7DGM6XEvNVo8BK3Bmjz/WynWUEI16EpF/AccDfUVkJ3AXEA2gqouAt4DTgWzABfwyFPdtSXZ2Nqu/+x53Qlp736rDCEpv8TQ4V+aqZOXWPL/lY6glURqe+6GojK1F1QHdLwIPqXJgFRu320127h5cxAQXeCcS6WppXIcx3VuzyUNVI/z93N5UtcX1MbxNZb/uoHDquRPSqBh7ekfftl0l7FhGbEVB/XF537FU9PW/KWSlu4b4be8QWeskAJUISoafQG1sit/yjcW49kHuhw3OaVJ/KjKOPrjgO4H4DW+FO4QeIzc3l0WLFpGTk8PkyZO5+uqrD2robV5eHps3b2bMmDFYc3bbBNxhbrqfwkHHkFSwnuiqEioTB1DWe7TfctGV+0ks3EhNTDJVcX3xRMdRnjIi4MQBUB3fh5qYJKKrS+vPuVKGtfk1mJ5h3rx57NixA4CPP/6Y6Ohorr/++qCusWzZMubPn4/H4yEyMpIbb7yRadOmtUe4PUJQyUNExuAMi41r/Jiq2sewLsYTFUtx/8ktlomoraJv7odEeGoBp31y35Djg0ocAIiwb8iPSSzaTGRtJa7kIVT18t+/Yoyv4uLi+sRRZ+3atUFf5/nnn6/v43O73Tz//POWPNog0FV1xwP/Ag7Fmd/RmAK263w3FFu+pz5xgPOPH1+6k+qEvkFfyxMVR0n6+BBGZ3qCpKQk+vfv32DAxqhRo4K+TuPNzWyzs7YJtC/jGaAGmIUzz2NEo69D2iU6E3aeyKYd2u5o23XYdJyIiAhuvPFGMjIyADjssMO4/PLLg75O483NZs6cGZL4eqpAm60OBc5W1XfaMxjT+cSXNpxC44mIpjx1RDOlIbK6jMT9W8DjxpU6gpq4wIfyGtOcsWPH8pe//IXq6mpiYg5uhN6vfvUrhg4dyoYNGxg3bpztlNlGgSaPrwCbpt0NibuGCE817mj/CyPHle9ucBzhqUGbqbBGuKtI3/EBkW5n+G6vkm3sHXYStbHJoQ3a9FgHmzjAmVt02mmncdppp4Uwop4r0ORxBfAvEXHh7N/RZGqxqrpCGVhnsmvXLiJdxd1uaGYsNSRQgwjUaARlxKKNurQ81BLpc8qtQvym/+Cv6yuGWiLlwLwPUQ/JOR9Q0YXncrQk0lXArl21rRc0AVu/fj2vvPIKNTU1nHHGGUydOjXcIZlmBJo89gHbgL+2UMY6zLuISNzEUEscbsSbA6LFQ5zWNHmjdxFDolYRKYpHoZwY/I+Z8L/hS+NkZExz8vPzufPOO6mudj6ArFmzhnnz5jFmzJgwR2b8CTR5/B04GvhfnJndgU0r7iYyMjLYUxXVLSYJxpfkkvzDl/7f0hPTqRh8bJPTZapE1pQ7HeXS/BiLCvUQs/NjYl35ANTEJLN/6PGon0737iB+w1tkZPQPdxjdxtdff12fOMBZQueLL75g9OjRfPrpp2zZsoUJEyYweXLLw8tNxwg0eZwAXK6q/2zPYEz7Syza1GxdoDKxmW1RRHDHBDCbVyLYN3g6MRX7EHVTldCvxWRjjC9/66r179+fp556ijfeeAOAl19+mSuuuML25ugEAv2fvQ1nHSnTzShQHZNMcfoEXKkhGHEtQnVCujMB0BKHCcLkyZM58cQT64+POOIIjjvuOP7zn/80KLdkyZKODs34EWjN47fAPSKyWlW3tWM8pp2Vpo0lbffn9bWP8tSRrc4yN6YjfP/999TU1HDMMccwY8YMpk6dSk1NDZGRkdTU1NSXi4qyVZU6g0D/Fe7BGaq7SUS24X+0lQ2L6AIqkzLYO/xk4sr3OGtVBbFESERNJQkl20AicCUPxRPVZJUaYw7Ktm3buP3226mtdUavrVq1ioULF9KnTx/OPvts/vGPfwAgIpx33nlBX3/z5s2sWbOGyspKhgwZwlFHHWWbmbVRoMnjO++X6QZqY1MoC3JtqpiyPPru+gTxjqlKLNzI3uEnWwIxIfHJJ5/UJw6AiooKvvzyS04//XTOP/98xo8fz5YtWxg/fjzDhw8P6tpvvfUWixYtanBu+PDhPPTQQ5ZA2iDQPczbfc8M07ml7fmqPnEARLqriC/ZQXma/5V4jQlGWlrTvXJ8z40bN+6gt5RdvHhxk3Pbtm3jyy+/ZPr06Qd1TRPkNrQAItJHREaJSJ/2CMh0PuKpJdLtZ9tYsTkcJjRmzJjB2LFj64+zsrI48sgjQ3Jt3xqNL99+FBO8gHueROR84G5gtM+5TcCdqvpi6EMznYVGRFEdm0pM1YGuLo9EUpFsK9aY0IiLi2PevHls2rSJqKgoRo4cGbJrn3nmmfzznw1nGfTr14+jj+66G5F1BoEuyX4h8A/gbeAPQB7QHzgfWCwikaratG5ouo3CjKNJ3bOSmIpCamOSKBz0IzyR1l5sQkdEGDNmDPn5+Tz99NOUlpZy0kkncfjhh7fpuhdccAGHHHIIK1euxOVyccghhzBjxgwSEmx16LYItObxO+AJVb2q0fm/isgi4HbAkkc35o7uRcEQax827auqqopbbrmFffv2AfDhhx/yP//zP21OIFOnTm1xnawNGzbw5ZdfMmjQII4//niio6PbdL+eINDkkQnc0MxjLwOXhiQaY0yPtnr16vrEAc4SJcuWLWtz8mjJZ599xrx581B1BoR8/fXX3Hbbbe12v+4i0A7zPCCrmceyvI8bY0ybpKQ0HULu71wovfnmm/WJA+CLL74gPz+/Xe/ZHQRa83gWuFtEIoGXcJJFP+BcnCarP7RPeKZdqJJUsJ5exTl4ImIoST+cysRB4Y7KGMaOHcu0adP45JNPAGe9q/Zex6rxjPWIiAgiI22R8NYEmjzuBaKBW3Fmm9epwFlp994Qx2XaUULJdpILvgcgkkrSdn3OnpGn44mKD3NkxsDNN9/MWWedRWlpKePHj2/3/oezzz6b7777rn5I78knn+x33olpKNBJgh7gdyLyv8B4YADwA/Cdqha1Y3ydRqSrsNtsBpVAVYMtOQQlKfs9qgP4c4ioLAHAE2e7A0a6CnEGHZpQGzVqVIfda+LEiSxcuJCVK1cyaNAgJk2a1GH37sqCWmHMmyg+aqdYOq3MzMxwhxBSJSUlFBQUNDg3dki/gD7hZWeXApB5iL1pQv9u97fRE7lcLt599122bNnCpEmTmDBhgjVbBSCYSYITcIbsZgGDgaNV9RsR+T3wiaq+3U4xht3cuXPDHUJI1dbWsnDhQpYtW0ZCQgIXX3wxp556akDPve666wCYP39+e4Zoepja2lrWrl1LYmJig1pHfn4+jz76KOvWrWPMmDHMnTuXQYNC2z/38MMP89VXXwHOgozFxcVceumlIb1HdxToJMHTgCXAZzhb0d7l83AVMBdnAqHpAqKiorj22mu5+uqrrXPQhN3+/fu59dZb2b17NwDHHnsss2fP5qWXXmL9+vWUlZUBsG7dOh555BEefPDBkN27oqKCr7/+usG5jz76yJJHAMR3iFqzhURWA1+r6uUiEoWzDW2Wt+ZxJrBIVTvtcJ2srCxdsWJFuMPoFqzm0fksWLCA7OzscIdx0IqKiti/v+EuDyJCc+9Nw4cPR5pZV63u9xBoc6KqsmPHDjweT/252NjYkNduwiUzM7NNLScislJV/U7TCLTZaixwk/fnxv+iJUDIhiaIyKnAfCASeEpVH2j0+KXAQ8Au76nHVPWpUN3fmK4mOzubzetWMTTRHe5QDkq1pxfQcGn/5hJHNDVU71jZ7LViapypa1XbA/+wmKwx7CcREAQPidV7qdq+O+Dnd1Y7ytq3RSHQ5LEXaG6P0sOAHaEIxjuP5M/AycBO4GsRWaKq6xsV/T9VvSYU9+ypysrKqK2tJTU1NdyhmBAYmujmtikl4Q7joGwpreKRjQNR7xDAhMhaXO7Gb03KiF5V/HxEPklRHtwKSdGephc7SGU1heypjGFIQhWxka23xnQF93/TviMiA00ei4F7RWQ98Ln3nIrIaOAW4OkQxTMVyFbVrQAishiYDTROHqYNnn32Wd544w3cbjfTpk3jhhtusK09TdiMTKri2jF7+M/uFGpVOGVgMd/uT+DT/CQUYWxyBVeMzCMmUnl9Z28+yEvBozCpdzmXjMgnKuiNJZpKjPaQGV3Z9gv1IIG+Y9wBjAOWA3u8517Hme/xLnB/iOLJAHJ9jncCP/JT7mwRmQ5sAm5Q1Vw/ZYwf69at49VXX60//vjjj5k4cSIzZ84MY1Smp/t2fwIbS51VbhdujmN8iou7x+fiVqFfnDN5b2tZLO/tOVBTXlWUyJjkSqall7Z47e3lMby4ow/5VdFMTHVxzpACYrpJ7SKcAp0kWAXMEpETgROBvkAhsFRV32vH+Px5A/iXqlaJyJXA88CMxoVE5ArgCoChQ23fiTq5uU3zrL9zxnSUCrfw0V7fJhZhbXEvxuyv5Pj+B5ri9lQ0nYf0WX4iBVVRHJteSt/Ypps+uT3weHZ/Smqct7rP9iWREOXmp4N7xNzmdhXoUN2hwA+quhRY2uixKGCQqoai32MXMMTneDAHOsYBUFXf2W1PAX7H7anqE8AT4Iy2CkFs3cKkSZOIiopqsLtaVlZza14a0/5UBY+f/6GbSuMaJI+xyRVEiVKrB0Za7XDFscMVxyf5SfzusF2kxjQcNJBXFV2fOOpsLokHinh7dyqf7UsiPtLDTzIKGZ9aEdLX1d0F2lqYA0xu5rGJ3sdD4WtglIiMEJEY4AKc+SX1RGSgz+GZwPchunePMGDAAG6//XbGjRtHZmYm1113HRMnTgx3WKYHS4jykJVW1uT8sF4Ntz5Oi3VzZeYeMhMrSIhsmCQq3JEs3ZPCE9n9+OP3A/kkPwmAvrG1TcoO7VXFVwW9+Pfu3hRVR7G7IoantvRnf7XNdwpGoH0eLW1WHYczUbDNVLVWRK4B3sEZqvuMqq4TkXuBFaq6BLjWO7ekFqfp7NJQ3LsnmTJlClOmTAl3GMbUu3jEPhKiPHyen0SNCkeklTOjf9PRY4emVHJoyh7u+nYwLnfDN/tP85OoVufzcE55HLERHo7sU86lh+SzeHsfiqqjGJdSwayMIl7NbTi7wK1CdlkcWWnl7fciu5lmk4d3ORLfFcJOF5GxjYrFAefhdFyHhKq+BbzV6NydPj//P+D/hep+xpjwixA4d2ghZw8pxKO0OoIqPbaGgmrfPhCtTxx1vt2fwJF9yhmXUsE943dSqxDtLTKkVxVfFCTVlxWUIQkh+QzcY7RU8ziLA8uQKHBnM+VygCtDGZQxpmeKEOerNadn7Gfrxrj6hHFMn1I+L0iqnysCkO7TgS4C0T7Xnda3lNzyWL4qTCQ2wsNPMoroH9e0w900r6XkcT/OXh2CM4t8Bk6fhK9qVa1pp9iMMcavQxKruHdCLptK4+kXV8PghGrSYt28/UOqd3hvNZGi/FARzcD4pm9RkRHw8xH7OH9YARGiRAaQsExDzSYPb1Ko+62HYBqOMcaETmK0hyk+fRSnDtrPtH4l/COnL2uLe/H2DzG8syeVqzPzODTF/0iq6AgbiHmwglmSPRWneWoazlpWhcDHwBOqur+l5xpjTEeo8ThzROp4VFial9wkeVR7BLcH4qMseRysQOd5jMSZXZ4OfIqzllV/nO1nrxGRE1R1S7tFaYwxzaj2CNGiiIC/9RRVG7ZJvfNDCu/8kEqNR8hKK+e0QUXsdMUyvFcVaX4mGhr/Aq15/AkoAn6kqvWT9kQkA2dk1MM4a1AZY8xB214ew1u7e1NWE8lRfUs5rl/zS4+U1kTw3NZ+bCyNp09MDT8bvo8xyZVM6V3GN0WJAESgzBhQXP+cXFcMb+w6MEz368JEVhT2QhEiUC4ekc+RfWy4biACTR7HA5f4Jg4AVd3lnYPxbKgDM8b0LBW1wmObBlDhnb+xfUcs8VGeZudevLYzjY2l8QAUVEfz3NZ+3DdhB5ccks/EIhf5lVGMT3WRkXCgw3yXK6bJdepGaHkQluxKs+QRoECTh+JM2vMngqZ7fBhjTFCyy+LqE0edtfsTmk0eO1yxDY5LayMpqo4iPa6WI5p5zuikSiJFcav/4VWuWhsbFKhAf1PLgPtEZJjvSe/xvTRa78oYY4LVL67pkNoBfs7VGZ3UsBO8T0wNfVrps/AAIxMriI3wkBxVy6jEhtc4pm/LK/SaAwKtedyAkyA2i8g3QB7QDzgCZwn137RPeMaYnqJ/XC2zBhXynx9SqdUIRidVcHy/4mbLn5lRRJU7gu+KE+gfV8O5QwtanGDoVliwcUD9zPQqTwRn9Ckiq085W8tiOSSxiqMteQQs0CXZc7xLk8wBjgQG4mzQ9CzwnKpWt1+Ixpie4tRBxUzvV0qlO6LJyKd9VVGs3Z9An9haDk9xERup/HzEvoCvvaM8ttGSJrCmqBf/PTqPY1vZE8Q01WryEJE4nJVt71fVRcCido/KGNNjJUR5SIhquMVsdmksj20aQK13OZIj0sr45SH5QV23d0wtESgenyVMWmvmMs1rtc9DVStxahu2XrEJmMfjYd26dWRnZ4c7FNMNLM1LqU8cACsLE9lXFdzWyakxbmZlFBEhzvieAXHVzBxo85sPVqC//SXAT7GOcROA8vJybrvtNnJynG1epk6dym233UZEhI1kMQfH42d0lPsgxnjOHFjMUX1LKamJYlB8dUCLMBr/Ak0e7wAPeTdiegunw7zBP513KXVjePfdd+sTB8BXX33F6tWrbQ8Rc9B+3K+E74vj65ucDktxHfQquMnRHpKjrZu2rQJNHn/3fv8v71djLc0DMT3M/v1NmwKKimzPaHPwxqVUcPO43awpcjrM/e08aDpWoMljRLtGYbqV6dOns2TJEtxuZ/vPxMREjjzyyDBHZbq6wQnVDE44UGNwK+ypiKZvbC2xkTZPuaMFOlR3e3sHYrqPkSNH8vvf/563336buLg4zjzzTJKTk8MdlulGdrmi+cvmAeyviSIuwsNFI/KZ3NsV1DXcCsvyUvi+OJ6MhGpOHbi/ySgv07yghiuIyExgKs48jx+AL1X1vfYIzHRt48aNY9y4ceEOw3RTr+T2YX+N8/ZV6Ynghe19mZC6I6hNnd7c1Zv39qQCsLE0nh8qovn16Lz2CLdbCnRJ9kHAqzhDdvd6v/oB94rICuCsxosmGgNQVVXFqlWrSExM5LDDDkPEhreYtms8TLe0NpJKdwS9gqg5fFPYq8Hx9yUJuGojrPYRoEBrHk/g1DamqepndX7nb9EAACAASURBVCdF5FjgX8DjwKzQh2e6soKCAm6++Wby853JXFlZWdxxxx2WQEybTejtYlleSv1xZmJFUIkDnEmDvjPOe0W5iY2wxBEoUX+7pzQuJOIC5qjqYj+P/Qx4UlV7NX1m55CVlaUrVqwIdxhttmDBgrBPuqu7f2ZmZqtlCwsLKS5uuDbRgAEDiI+PD0ksmZmZzJ07NyTX6srOOeccyov2MizJHe5QOowqlGo8VUQTjZskcRHpnfxXqxEUay+qiSKWWlKkrP4x3+eXaALlxAECKKlSRoJ0nyG820sj6dW7Hy+99NJBX0NEVqpqlr/HAq155AH+NwF2zge+wIzp0oJ54/d4mn6K83fOmGCJQLJU4O9tab8mUo1To6gkBtVE+kjDtauKtRcu4uqPe1HRrRJHRwg0edyPt3+j0U6Cg4G7gd+3Q2ymka72KXvTpk3ceuut1NY6k7nS09P585//TGxsbCvPNMHIyMigqvYHbptSEu5Qwk4V5q7s0+CcJyKqwe+m1gM3rkprMM05MjqG2yZ2r87y+79JJjYjo92uH2jymAn0AbZ6l2Sv6zCfAuQDJ4nISd6yqqrnhzxS0+WMHj2aBx54gKVLl5KYmMjpp59uicO0KxEYklBFrs9GUUMTGtYoIgTiIj2U1x6Y1+yqjeAvm/tzxqAihvayGkggAk0efYHN3i+AZKASqOs8Tw9xXKabGD16NKNHjw53GKabqvYIeyuj6R9XTbR36bSLR+Tz/NZ0dlXEMiShip8Nb7j6boQ4e4Es3t6nfgvaWo1gXXECOWWx3Dch1yYdBiDQSYIntHcgxhgTjO+L43h2az9c7kgSo9xcPjKPkUlVDIqv4f8dtptqjxAT4T8JHJteyuikCp7PSWdb+YG+D5c7ki1lcYxLaa6L19SxZU6NMV2OKvWJA6CsNpIXcxv2dfgmjmq3sHxvEq/m9mZbmdOklR5Xy5jkygbPEdTvdrimqeAWxDemkdzcXObPn092djaHHXYY119/Penp1opp2tcn+Un1iaNOfkV0M6Vh4eb+ZJc5IwU/yEvhqlF5HJZSwYn9i9laFsvm0niixcMZGUX0tQ2iAtLpah4icqqIbBSRbBG51c/jsSLyf97HvxSR4R0fpanz8MMPs2nTJjweD2vXrmXhwoXhDsn0ABtKmg4ZH55Y6aeksw5WXeIAUISle1L4qiCR/TWRXDdmD/dN2MH9k3Zw0gAbsRaoTlXzEJFI4M/AycBO4GsRWaKq632KXQYUqWqmiFwAzANsdFcY1NTUsGXLlgbnNmzYEKZoTE8yOKGaNfsPzEsWlHOHFjQoU+MRPshLJrs0rvHT2Vwax6ZSJ6FkJlYwrFcV09JLiY+0WkegOlvNYyqQrapbVbUaWAzMblRmNvC89+eXgBPF1rsIi+jo6CYzzQ899NAwRWN6khn9izk8pRxBiY90c97QAgbEN3zj//u2vryxK43vSxIanBe0fpQVQHZZPEvzUnno+0EUV9u2RIHqbMkjA8j1Od7pPee3jKrWAsU4c1AaEJErRGSFiKyoW1vJhN5vfvMbDj30UKKiopg8eTK//vWvwx2S6QFiI5WrRu1l3qQd3D9xB8f1aziDvMotrGq08GFCpJufDcsnNcZ/7cLljmRVUaddZanTabHZSkQSgLOBQcBG4A1VdTcqcwhwu6rOabcoD4KqPoGzoCNZWVk2aLudDB48mHnz5oU7DNNDNV4Bt9YD/7ejL1/tS6Txf/q02FqOSS/D5Y7ktZ1pfq8XF2nL5wSq2eTh3a/8U2A44AISgI0icrGq+q4ymA5cAoQieewChvgcD/ae81dmp4hEASlAAcaYHm/53mQ+35fkc0YBIVo8DIir5uO9SRyXXkKfmBrWFSfwfUk8xd59QYYkVDGld3lY4u6KWqp5/AGoAsao6mYRmQDMBz4SkV+o6sEv1di8r4FRIjICJ0lcAPysUZklOMnqc+Ac4AMNZGlgY0yn5ezql8ymkngGJ1Qzc+B+4g5ilrfvhD+HcFSfEr4qSGJFofP1ZUEiN479gclpLtweWF8SjwCHplQEtZlUT9dSn8cM4C5V3Qygqt96zy0AFovIDaEOxtuHcQ3wDvA98IKqrhORe0XkTG+xp4E+IpIN/AZoMpzXGNO1LNnZm9d29mF9SQLv7knlrznpVHuEj/Ym8Vpub3LKAlsTbWSj4bpR4qHCHYnHp4N8W3kc28qd60VGwPjUCg5PtcQRrJZqHr1xlmKv5/2Ef4uIbAce9a6q+2IoA1LVt4C3Gp270+fnSuDcUN7TGBNeKwoTGxyv3Z/Awk0HJvYtzUvhysw8Dk9tedmQ4/qVUFAVxZcFiSRFu/np4EJW++kEj2q0v0dpTQRv7EpjhyuG0UmVzBpURIytb9WilpLHFpyhs8sbP6CqC0UkD/g7YOteGWPapHdMbX3fA0B8pKfJxL6P85NbTR6RAmcPLeTsoYU+13azuqgXVR6noeWwFBdDGq2c++SW/mwtc5q8drpiKauJ4BeH2DZFLWmp2eo94HIR8VtGVV8GTgMOaY/AjDE9x08HFxIf6QzkjBIPpw0salImuplFDlszOKGaOw7fyXlD93H5yDyuyGy4b0e1W+oTR51vbMhuq1qqefwR+BBIBPzO2VfVD0XkKOBHoQ/NGNNTZCZV8T8TctnhimVgXDWJ0R5yK2L5qsAZORUT4eGk/sWtXKV5qTFupjeaCwLgUVi3v+lSJx61DpDWNJs8VHUP8O/WLqCqGwBbk8IY0yaxkcqopAMd3hcP38eRaWUUVEdzeIqL1JjQ7tHuUXhs04D6ZUp8Devlf50sc0CnWtvKGGPqiMChKZXUeiqp8ghv705lTVEC6XE1nJlRRHqc/5niBVVRJES6iY9quZlrY0mc/8SRUMkl1t/RqpYmCQaV5lXVFoUxxoTUJ/lJvL6zNxXuCPAOt91ZEcsuVwy3H76LCJ/WpfLaCB7Pdjq+o8XDrIwiTmxhldxqT9Pu3OPSSzh/mM05DkRLNQ8BSnEm5X3RMeEYYw7GjrJI7v8mOdxhhJRbI8jTVKBp/8Peqhju+6Y30T6fcUs8CZThdHzXaASv7kzj010RRIn/JUdUIRI3buo+9yob9nm4v6B7/B53lEUyqh2v31LyuBi4EDgPOBZnhdvF3smCxphOovHKxt2Fy+WCvDy/j4kICUMnEhFxoPbgycsDl8u3FBH9DyU23mma8ng8VFQ4Q30TEhIQETLcbkpKSvB4PCQlJRET030Gj46iff82pLWVPUSkN84yIBcAP8ZZIPH/cBLJpnaLLISysrJ0xYoVrRc0xnQaZWVlzJkzh8rKhp3X8fHxXHnllcyYMaPB+ffee48FCxbUHycnJ/P0008TGxuLy+Xit7/9Lbm5zqLdw4YN48EHHyQ+vmmfhzlARFaqapa/x1pdkl1Vi1T1SVU9EWehwkXAKcB6EflzaEM1xhhHYmIit99+O2PGjKF///5ceOGFLFq0iOeff75J4gA4+eSTSUtLIyYmhqlTp3LvvfcSG+ssQ7J8+fL6xAGwfft2Pv744w57Ld1RsKOt3ICHuqUq/TVGGmNMiEyYMIGHHnoo4PIpKSmkpKRw++23NzjfuPYCUFVV1eb4erJWax4ikiIic0TkfWA3zsKF7wOHqep/t3eAxhjTVtOnTycx8cD6WUlJSUybNi2MEXV9LQ3V/RlOP8dMnKTxAnCTqq7uoNiMMSYk+vTpw8MPP8z7778PwMyZM+ndu3eYo+raWmq2+jvOUN2XcPbOUOAYETnGT1lV1b+0Q3zGGBMSAwYM4Oc//3m4w+g2WuvzSMLZjKnxhkyNKWDJwxjTbnJycnj//fdJSEjgtNNOIy0tjZdffpl33nmHXr16cdFFF7Xp+mvWrOHFF1+ksrKSM844gxNOsAXDW9LS2lat9ocYY0xHyMnJ4aabbqKmpgaApUuXctFFF/H888/Xl7n//vsZOHAgUVHBr7qUn5/PvffeW3/9TZs2kZaWxsSJE0PzArohSxDGmE5v6dKl9W/sAPv27eODDz5oUKa2trbJqCqXy8X777/P0qVLnUmHzVi1alWD6wN89dVXIYi8+2o2eYjIQBF5WUROaaHMKd4y/donPGOMcWaENzZ48OAm52JiYup/Lisr4/rrr+fRRx9l/vz5/OY3v6G8vNzv9f1dy985c0BLNY+bcDZ6ereFMu8CI4AbQxmUMcb4OvXUU0lPT68/njBhAnPmzOGEE04gIiKChIQELrvssgbJY/ny5ezZs6f+ePfu3c1ODBw3bhw//elPiYx01rk66qijOOmkk9rp1XQPLTUOzgIe1hbWL1FVFZHHgRuAW0IdnDHGAKSlpbFw4UJWrFhBQkICEyc661rdcMMNXHXVVURFRREdHd2gKau2tumS7f7O1ZkzZw7nnHMONTU19OnTp11eR3fSUs1jGLA+gGt8DwwPSTTGGNOM2NhYjj32WCZPntxgQcT4+Hiio6OblJ8+fTopKSn1x71792b69Okt3iM5OdkSR4BaqnlUAIGsTZzoLWuMMZ1G7969eeSRR1i2bBkiwowZM0hOPvCW5na72bp1K/369WuQZExgWkoe3wBn0vpWtLO9ZY0xplPp06cP55xzTpPzu3bt4q677mLv3r1ERUVx2WWXccYZZ4Qhwq6rpWarhcBlInJJcwVE5BfAL4HHQh2YMcYcDFXltdde46abbuKhhx5i165dTcr84x//YO/evYDTD/Lss89SVlbW0aF2aS1NEnxZROYDz4rINcB/gB04s8mH4izLngX8SVVf7YhgjTGmNSUlJTzzzDOAM9lv48aNPP744/UjqYD6xFGnurqa/fv3N1g80bSsxUmCqnojTrNUCc7Q3ceBJ4Df4qx7NVtVb2rvII0xJlCNJwPu3buXrVu3Njh37LHHNjgeMWKEzesIUqvz+FX1DeANEYkC6oYhFKhq82PejDGmgxUXF1NYWNhkOG5UVFSDOSIAP/3pT4mOjubzzz8nIyOD888/vyND7RYCXgTGmyz8byhsjDFhpKrccccdFBcXNzgfExPDpZdeSmpqaoPzIsKsWbOYNWtWR4bZrbS0n8edQVxHVfW+tgQiImk4e6MPB7YB56lqkZ9ybmCt93CHqp7ZlvsaY7q+LVu2sG3btgbnxo0bx+233279GO2kpZrH3TjzN8ppfbtZBdqUPIBbgaWq+oCI3Oo99jdrvUJVJ7XxXsaYbiQ5ORkRwXdBjP79+1viaEctdZhvAaKBlXjXuVLV9Ga+QrEw4mygbn3l54GfhuCaxpgeoF+/fsyePbv+ODU1lXPPPTeMEXV/zSYPVR0FHAOsw6lV5InIKyJyrojEt0Ms/VX1B+/Pe4D+zZSLE5EVIvKFiFiCMaYb2rp1K6+//jrfffddwM+ZM2cOGRkZ9O/fnyeffNJGT7WzFjvMVXUFsAK4SUSm4+xp/hjwjIgsAR5X1Y8CvZmIvA8M8PPQ7xrdV0WkuQUZh6nqLhE5BPhARNaq6hY/97oCuAJg6NChgYZojAmzZcuW8cgjj9Q3Qf3sZz/jggsuCOi5MTExxMTEEBsb254hGoLYDEpVP1LV/waGAIuA84Hrg7mZqp6kqof7+Xodp2YzEJy9RIC9zVxjl/f7VuBDYHIz5Z5Q1SxVzWo8TM8Y03m98MILDfouXnnlFWpqaqiqqsLtdrfbfXfs2MHy5cspKChot3t0JwEP1RWRY3FqHufg7G3+EqHdt3wJcAnwgPf7635i6A24VLVKRPoCxwIPhjAGY0yYNU4Qbrebhx9+mM8//5xevXpxySWXMHPmzJDe89VXX+XZZ58FnHkhv/vd7zjiiCNCeo/upsWah4hMEZEHRWQ7sBSn1nED0E9VL1DV5SGM5QHgZBHZDJzkPUZEskTkKW+ZQ4EVIrIGWAY8oKqBLBtvjOkifDu+AUaNGsWnn36Kx+OhtLSUhQsXkp+fH7L7VVdXs3jx4vrj2tpa/vnPf4bs+t1VS/M8NuLsEvgBcBfwiqqWtFcgqloAnOjn/ArgV96fPwPGt1cMxpjwO+OMMxg8eDDffvstI0eObLL7n8fjIScnp8ms8YNVW1tLVVVVg3PNbVdrDmip2WoUUAkcAUwBHhRpfrpHiIbrGmMMEydOZOLEiYCz0OGnn35a/1hMTAxjx44N2b0SEhI47rjjWL78QEPKKaecErLrd1ctJY97OiwKY4xpxsyZMykoKGDp0qWkpqZy8cUXN9jUKRSuvfZaxowZQ05ODpMnT2batGkhvX531NKS7JY8jDFhUVBQwKOPPlrfdHXttddy0UUXtdv9oqOjbZ2rIAU8VNcYYzrKokWLWLVqFW63m02bNvHQQw+FOyTTiCUPEzI1NTVs2LCBkpJ2G1dheogNGzY0ON6+fTsVFRVhisb4E/A8D2NakpOTwz333ENhYSFRUVFcddVVIR+Lb3qOsWPH8uWXX9YfDx8+nPj49lgVyRwsq3mYkPjb3/5GYWEh4Ax9fPrpp5sMfzQmUFdddRWTJ08mMjKS0aNHc9NNB7dh6Y4dO/jwww/Zt29fiCM0VvMwIdF4T+iKigrKyspsjSFzUPr06cM997RtzE7jWeO33XYbWVlZoQjPYDUPEyKNhzaOHTuWPn36NFPamPZVU1Njs8bbmdU8TEicd955xMfHs2LFCoYOHcp5550X7pBMD1a3kKIvmzUeWpY8TEhEREQwe/bsJusSGRMOCQkJTJ8+nQ8//LD+3Kmnnhq+gLohSx7GmG5p7ty5jB07lq1btzJp0iSbNR5iljyMMd1SdHQ0p59+erjD6Lasw9wYY0zQrOZhjOkyysrKWLx4MTk5OUycOJH/+q//IirK3sbCwX7rxpgu48EHH2T16tUArF27lvLycn75y1+GOaqeyZqtjDFdQllZWX3iqOO7z4fpWJY8jDFdQlxcXJN9PPr1sz3owsWShzGmS4iKiuLKK68kJiYGgJSUFGuyCiPr8zDGdBnHHXcckydPZvfu3YwYMYLo6Ohwh9RjWc3DGNOlJCYmMmzYMEscYWY1D2NMl5Gfn88f//hH1q9fT0ZGBtdddx1jx44Nd1g9ktU8jDFdxpNPPsn69esB2LVrFw8//DCqGuaoeiZLHsaYLiM7O7vB8Z49eygtLQ1TND2bJQ9jTJcxfvz4BsfDhw9vMnzXdAzr8zDGdBmXX345tbW1rF69mhEjRnDVVVeFO6Qey5KHMabLSExM5Le//W24wzBY8jDGGHbu3Mnjjz/Otm3bmDRpEldeeSWJiYnhDqtT6zR9HiJyroisExGPiDS7S72InCoiG0UkW0Ru7cgYjTHd07x581izZg3FxcUsX76cp59+OtwhdXqdqebxHfBfwOPNFRCRSODPwMnATuBrEVmiqus7JkRjTHewc+dO3nnnHSIiIjj22GPZvn17g8fXrl0bpsi6jk6TPFT1ewARaanYVCBbVbd6yy4GZgOWPIwxAcnLy+PGG2+koqICgPfee4++ffuyb9+++jIjR44MV3hdRqdptgpQBpDrc7zTe84YYwKyfPny+sQBzlLvU6ZMYeDAgQAceuih/OpXvwpXeF1Gh9Y8ROR9YICfh36nqq+H+F5XAFcADB06NJSXNsZ0Yb169WpyLjs7m0WLFlFVVUVcXFwYoup6OjR5qOpJbbzELmCIz/Fg7zl/93oCeAIgKyvL1i8wph0tWLCgyezvcKiL4brrrmu2TG1tbZNzOTk5XH/99SGLIzMzk7lz54bsep1RV2u2+hoYJSIjRCQGuABYEuaYjDGdRHx8PPHx8S2WiYqKarIir9U2gtdpOsxF5CxgAZAO/FtEVqvqKSIyCHhKVU9X1VoRuQZ4B4gEnlHVdWEM2xgDXe5T9o4dO5g/fz7Z2dkcfvjhXH/99aSnp4c7rC5FesKKlFlZWbpixYpwh2GMMV2KiKxUVb/z7rpas5UxxphOwJKHMcaYoFnyMMYYEzRLHsYYY4JmycMYY0zQLHkYY4wJWo8Yqisi+cD2VguaQPUF9rVaypjwsL/P0Bmmqn4nwPSI5GFCS0RWNDf225hws7/PjmHNVsYYY4JmycMYY0zQLHmYg/FEuAMwpgX299kBrM/DGGNM0KzmYYwxJmiWPIwxxgTNkkc3JyIqIn/3OY4SkXwRebOV56WKyH8HcP0G5URkkIi81LaoTU8gIn8Sket9jt8Rkad8jv8oIneKyK3e47tF5CY/1xkuIt95f54kIqd3RPw9nSWP7q8cOFxE6rZXO5lmtu5tJBVoNXk0Lqequ1X1nKCjND3Rp8AxACISgTO57zCfx48B3lXVB4K45iTAkkcHsOTRM7wFnOH9+ULgX3UPNP40JyLfichw4AFgpIisFpGHRCRRRJaKyDcislZEZnuf0ric76fAOBF51lt+lYic4D1/qYi8IiL/EZHNIvJgu/8GTGf0GXC09+fDgO+AUhHpLSKxwKHABBF5rPETReQIEVkjImuAX3vPxQD3Aud7/x7P9/59pXsfjxCRbBFJF5HnRGSRiKwQkU0iMstbJtL7d/y1iHwrIle2+2+hi+o029CadrUYuNPbVDUBeAY4rpXn3AocrqqTwGnuAs5S1RIR6Qt8ISJL/JQb7nONXwOqquNFZCzwroiM9j42CZgMVAEbRWSBquaG4LWaLkJVd4tIrYgMxallfA5k4CSUYmAtUN3M058FrlHVj0TkIe/1qkXkTiBLVa8B8P7dXQQ8ApwErFHVfBEBGA5MBUYCy0QkE/gFUKyqR3oT2Kci8q6q5rTDr6BLs5pHD6Cq3+L8R7kQpxZyMAS4X0S+Bd7H+U/ev5XnTAP+7o1hA876YnXJY6mqFqtqJbAeGHaQcZmu7TOcxFGXPD73Of7U3xNEJBVIVdWPvKf+1sL1n8FJCABzcJJOnRdU1aOqm4GtwFhgJvALEVkNfAn0AUYdxOvq9qzm0XMsAf4XOB7nP0SdWhp+iIhr5vkXAenAEapaIyLbWigbiCqfn93Y32JPVdfvMR6n2SoXuBEowXmjT2vLxVU1V0TyRGQGTi3jIt+HGxfH+ZA0V1Xfact9ewKrefQczwD3qOraRue3AVMARGQKMMJ7vhRI8imXAuz1Jo4TOFBTaFzO18d4/7N6m6uGAhvb9jJMN/MZMAsoVFW3qhbiDMI42vtYE6q6H9gvItO8p3wTgr+/x6dwasAvqqrb5/y53n6QkcAhOH+b7wBXi0g0OH+3ItKrTa+wm7Lk0UOo6k5VfdTPQy8DaSKyDrgG2OQtX4DT3vudt035H0CWiKzFaQbY0Ew5XwuBCO9z/g+4VFWrMOaAtTijrL5odK5YVVtaVv2XwJ+9zUvic34ZMK6uw9x7bgmQSMMmK4AdwFfA28BV3ibUp3CaUb/xDvx4HKsV+2XLkxhjujURyQL+pKrH+Zx7DnhTVW1O0kGyjGqM6ba8EwyvpmHTlgkBq3kYY4wJmvV5GGOMCZolD2OMMUGz5GGMMSZoljyMMcYEzZKHMcaYoFnyMMYYEzRLHsYYY4JmycMYY0zQLHkYY4wJmiUPY4wxQbPkYYwxJmiWPIwxxgTNkocxxpigWfIwxhgTNEsexhhjgmbJwxhjTNAseRhjjAmaJQ9jjDFBs+RhjDEmaJY8jDHGBM2Sh+k0REQD+Dre++V7rkhEPhGRE32uNbyZ5y9udM97RGStiJSISKmIrBCR8zv+1TckIjNF5PoQXCdGRO4WkUkH+fybReT4tsZhup+ocAdgjI+jfX6OBz4A/gf4t8/59cAU788XAVuBPsANwH9E5EhVXe1T/ibgU5/jfY3umQw8572uGzgHWCwiblV9qU2vpm1memN5pI3XiQHuArYBq1su6tfNwGPAh22Mw3QzljxMp6GqX9T9LCKJ3h+3+J73Plb347eq+p333HIgF7gc+LVP8Y2Nn9/onjc0OvWuiBwG/AIIZ/IwplOzZivTLahqGbAJGB6CyxXgfGJvlog8523iOkNE1ouIS0T+LSJpIpIpIstEpNxbZoLP8+qa02b5u57357uBG4FhPs1tz7UQy5kistJ7vyIR+VJEfux9uNT7/Vmfaw33Pu8Bb5NdmYjsFJF/iMgAn+tuw6nV3dWo2bDV1+A9HiwiL4jIXhGpEJEtInJfS79X03VYzcN0CyISCQwBvmv00LMikgbsBf4F/E5VK/w8PwpIBM7AaTK6IIDbDgXuBW4HEoAFwBM4CexJ4EHgDzjNYIepqgb4cp4CRgEzgLO85/L9FRSRkTg1pPnAb4E44AggzVtkBk2b/37wfu8H3A/sBtJxEtYHInK4qnq8917mvf5T3ues97l2a/6K0/x4BbAfOAQYG+BzTSdnycN0ZZHeN/004HfAQOBV72NVwJ+Bd4ES4HjgFmAkMNv3IiJyFPC597AWuEZVXwvg/mnA0aq6xXudCThv4Jeo6l+95wTnTXss8H0gL0pVd4rID0BVS01uXpOBUlX9rc+5t3x+/tr7vUnzn6rOqfvZm3w/B3YC04CPVHWViNQCOxs1KQaaPKYCF6rqG97jDwN8nukCLHmYrsy3A7gcuEVV3wRQ1R+Aa3we/1BE8oCFIjJRVdf4PLYWOBJIxal5PCYiJar6r1buv60ucXhle79/4OdcBgEmjyCtBVJE5HngH8CnqloeyBNF5DTgDuAwnIEDdUYDH4UgttXAH0SkD/CBqu4IwTVNJ2F9HqYruwDnTT8TSFXVB1spX9cBfoTvSVUtV9UVqvq+twP9b8C8AO6/v9FxtZ/zdefiArhe0FR1I05N6hCcGsc+EfmniKS39DwRORJYglPTuBhnpNtRIY71fGAF8Cdgu4is9h1Obbo2Sx6mK1vnfdPfoqq1AZTXRt+b8w0wxNskFmqV3u+NO+R7H+wFVfXfqnocTuf2ZcBJOP0vLTkLpx/lfFVd4m2W2hPgLQN6Daq6S1Uv9cZ1tPf6S7w1EdPFWfIwPck53u8rWyl3LE47fyAJKVh7gRrg0LoT3mHJxzQqV02QNQBVI1MDKAAAAbZJREFULVbVf+L0+4zzuQ5+rhUP1DTqxL/Iz2X9xRHoa6iLy+NNTvfgDCwY1vqrMZ2d9XmYbsk73DUJZ4JgCTAdpzP7FVX91ltmGPAMsBjYgjPa6iyc5rCr2yMuVfWIyOvADSKyHaeJ60ag8QiwDUB/EbkUZwTZPlXd1vh6InIlzqf6/+CMmhoFnIsz0glVrRaRHOA8EfkOp9bwLfAecL2IPAK8gfPG/3M/IW8AzhCR/wBlOPNmSlt7DSKSArzjjWMTEOsts4f26fsxHU1V7cu+Ot0Xzhu5Apf6eex472OHt/D8C3Da24txPj1n4wyrjfUpk4LTv5GD86a6B6ez+/QA4nsOWNHo3KXeuBJ9zg33npvlc64/8DpOUtuOM5S1wfVwPu0/i/MpX4HnmonjaJzRXLu9ryEHp7/G93XOxEkYld5rDfeevxlnYmU58D5O4lGc0WZ1zz0C+MJbRoHjA3kNOMniSWAj4MKZ2f8mMD7cf1v2FZov8f5DG2OMMQGzPg9jjDFBs+RhjDEmaJY8jDHm/7dXxwIAAAAAg/ytx7C/JGKTBwCbPADY5AHAJg8ANnkAsMkDgC1Q5GQMeus3PAAAAABJRU5ErkJggg==\n",
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ]
          },
          "metadata": {
            "tags": [],
            "needs_background": "light"
          }
        }
      ]
    }
  ]
}